// @yo-protocol/react/dist/index.cjs
"use client";
"use strict";
"use client";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  YieldProvider: () => YieldProvider,
  useAllowance: () => useAllowance,
  useApprove: () => useApprove,
  useClaimMerklRewards: () => useClaimMerklRewards,
  useCreateYoClient: () => useCreateYoClient,
  useDeposit: () => useDeposit,
  useGlobalVaultHistory: () => useGlobalVaultHistory,
  useLeaderboard: () => useLeaderboard,
  useMerklCampaigns: () => useMerklCampaigns,
  useMerklRewards: () => useMerklRewards,
  usePendingRedemptions: () => usePendingRedemptions,
  usePerformanceBenchmark: () => usePerformanceBenchmark,
  usePreviewDeposit: () => usePreviewDeposit,
  usePreviewRedeem: () => usePreviewRedeem,
  usePrices: () => usePrices,
  useRedeem: () => useRedeem,
  useShareBalance: () => useShareBalance,
  useSharePriceHistory: () => useSharePriceHistory,
  useTokenBalance: () => useTokenBalance,
  useTotalTvl: () => useTotalTvl,
  useUserBalances: () => useUserBalances,
  useUserHistory: () => useUserHistory,
  useUserPerformance: () => useUserPerformance,
  useUserPosition: () => useUserPosition,
  useUserPositions: () => useUserPositions,
  useUserRewards: () => useUserRewards,
  useUserSnapshots: () => useUserSnapshots,
  useVaultAllocations: () => useVaultAllocations,
  useVaultHistory: () => useVaultHistory,
  useVaultPendingRedeems: () => useVaultPendingRedeems,
  useVaultPercentile: () => useVaultPercentile,
  useVaultPerformance: () => useVaultPerformance,
  useVaultState: () => useVaultState,
  useVaultTransactionHistory: () => useVaultTransactionHistory,
  useVaults: () => useVaults,
  useYieldConfig: () => useYieldConfig,
  useYoClient: () => useYoClient
});
module.exports = __toCommonJS(index_exports);

// src/context.tsx
var import_core = require("@yo-protocol/core");
var import_react = require("react");
var import_wagmi = require("wagmi");
var import_jsx_runtime = require("react/jsx-runtime");
var YieldContext = (0, import_react.createContext)({
  ready: false,
  defaultSlippageBps: 50
});
function YieldProvider({
  children,
  partnerId,
  defaultSlippageBps = 50,
  onError
}) {
  const value = (0, import_react.useMemo)(
    () => ({ ready: true, partnerId, defaultSlippageBps, onError }),
    [partnerId, defaultSlippageBps, onError]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(YieldContext.Provider, { value, children });
}
function useYieldConfig() {
  const ctx = (0, import_react.useContext)(YieldContext);
  if (!ctx.ready) {
    throw new Error("useYieldConfig must be used within a YieldProvider");
  }
  return ctx;
}
function useYoClient() {
  const { ready, partnerId } = (0, import_react.useContext)(YieldContext);
  const chainId = (0, import_wagmi.useChainId)();
  const publicClient = (0, import_wagmi.usePublicClient)();
  return (0, import_react.useMemo)(() => {
    if (!ready || !(0, import_core.isSupportedChain)(chainId) || !publicClient) {
      return null;
    }
    return (0, import_core.createYoClient)({
      chainId,
      partnerId,
      publicClients: {
        [chainId]: publicClient
      }
    });
  }, [ready, chainId, publicClient, partnerId]);
}

// src/hooks/useVaultState.ts
var import_react_query = require("@tanstack/react-query");

// src/utils/vault.ts
var import_core2 = require("@yo-protocol/core");
function isVaultId(value) {
  return value in import_core2.VAULTS;
}
function resolveVaultAddress(vault) {
  if (isVaultId(vault)) {
    return import_core2.VAULTS[vault].address;
  }
  return vault;
}

// src/hooks/useVaultState.ts
function useVaultState(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query.useQuery)({
    queryKey: ["yo-vault-state", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultState(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { vaultState: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePreviewDeposit.ts
var import_react_query2 = require("@tanstack/react-query");
function usePreviewDeposit(vault, assets, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query2.useQuery)({
    queryKey: ["yo-preview-deposit", vaultAddress, assets?.toString(), client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.previewDeposit(vaultAddress, assets);
    },
    enabled: enabled && !!client && assets !== void 0 && assets > 0n,
    staleTime: 15e3
  });
  return { shares: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePreviewRedeem.ts
var import_react_query3 = require("@tanstack/react-query");
function usePreviewRedeem(vault, shares, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query3.useQuery)({
    queryKey: ["yo-preview-redeem", vaultAddress, shares?.toString(), client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.previewRedeem(vaultAddress, shares);
    },
    enabled: enabled && !!client && shares !== void 0 && shares > 0n,
    staleTime: 15e3
  });
  return { assets: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaults.ts
var import_react_query4 = require("@tanstack/react-query");
function useVaults(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error } = (0, import_react_query4.useQuery)({
    queryKey: ["yo-vaults"],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaults({ secondary: false });
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { vaults: data ?? [], isLoading, isError, error: error ?? null };
}

// src/hooks/useVaultHistory.ts
var import_react_query5 = require("@tanstack/react-query");
function useVaultHistory(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const yieldQuery = (0, import_react_query5.useQuery)({
    queryKey: ["yo-vault-yield-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultYieldHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  const tvlQuery = (0, import_react_query5.useQuery)({
    queryKey: ["yo-vault-tvl-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultTvlHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return {
    yieldHistory: yieldQuery.data ?? [],
    tvlHistory: tvlQuery.data ?? [],
    isLoading: yieldQuery.isLoading || tvlQuery.isLoading,
    isError: yieldQuery.isError || tvlQuery.isError,
    error: yieldQuery.error ?? tvlQuery.error ?? null,
    refetch: () => {
      yieldQuery.refetch();
      tvlQuery.refetch();
    }
  };
}

// src/hooks/useVaultTransactionHistory.ts
var import_react_query6 = require("@tanstack/react-query");
function useVaultTransactionHistory(vault, options) {
  const { limit, cursor, filters, allNetworks = false, enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query6.useQuery)({
    queryKey: ["yo-vault-tx-history", vaultAddress, allNetworks, cursor, limit, filters, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (allNetworks) {
        return client.getVaultHistoryAllNetworks(vaultAddress, { cursor, limit, filters });
      }
      return client.getVaultHistory(vaultAddress, limit);
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { history: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useGlobalVaultHistory.ts
var import_react_query7 = require("@tanstack/react-query");
function useGlobalVaultHistory(options) {
  const { cursor, limit, filters, enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = (0, import_react_query7.useQuery)({
    queryKey: ["yo-global-vault-history", cursor, limit, filters, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getGlobalVaultHistory({ cursor, limit, filters });
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { history: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPerformance.ts
var import_react_query8 = require("@tanstack/react-query");
function useVaultPerformance(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query8.useQuery)({
    queryKey: ["yo-vault-performance", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPerformance(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { performance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPercentile.ts
var import_react_query9 = require("@tanstack/react-query");
function useVaultPercentile(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query9.useQuery)({
    queryKey: ["yo-vault-percentile", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPercentile(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { percentile: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePerformanceBenchmark.ts
var import_react_query10 = require("@tanstack/react-query");
function usePerformanceBenchmark(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query10.useQuery)({
    queryKey: ["yo-performance-benchmark", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getPerformanceBenchmark(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { benchmark: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultAllocations.ts
var import_react_query11 = require("@tanstack/react-query");
function useVaultAllocations(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query11.useQuery)({
    queryKey: ["yo-vault-allocations", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultAllocationsTimeSeries(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { allocations: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPendingRedeems.ts
var import_react_query12 = require("@tanstack/react-query");
function useVaultPendingRedeems(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query12.useQuery)({
    queryKey: ["yo-vault-pending-redeems", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPendingRedeems(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { pendingRedeems: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useSharePriceHistory.ts
var import_react_query13 = require("@tanstack/react-query");
function useSharePriceHistory(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query13.useQuery)({
    queryKey: ["yo-share-price-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getSharePriceHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { history: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useTotalTvl.ts
var import_react_query14 = require("@tanstack/react-query");
function useTotalTvl(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = (0, import_react_query14.useQuery)({
    queryKey: ["yo-total-tvl", client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getTotalTvlTimeseries();
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { tvl: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePrices.ts
var import_react_query15 = require("@tanstack/react-query");
function usePrices(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = (0, import_react_query15.useQuery)({
    queryKey: ["yo-prices", client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getPrices();
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { prices: data ?? {}, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPosition.ts
var import_react_query16 = require("@tanstack/react-query");
var import_wagmi2 = require("wagmi");
function useUserPosition(vault, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi2.useAccount)();
  const userAddress = account ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query16.useQuery)({
    queryKey: ["yo-user-position", vaultAddress, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getUserPosition(vaultAddress, userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { position: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPositions.ts
var import_react_query17 = require("@tanstack/react-query");
var import_wagmi3 = require("wagmi");
function useUserPositions(account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi3.useAccount)();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query17.useQuery)({
    queryKey: ["yo-user-positions", userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getUserPositionsAllChains(userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { positions: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useTokenBalance.ts
var import_react_query18 = require("@tanstack/react-query");
var import_wagmi4 = require("wagmi");
function useTokenBalance(token, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi4.useAccount)();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query18.useQuery)({
    queryKey: ["yo-token-balance", token, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!token || !userAddress) throw new Error("Missing parameters");
      return client.getTokenBalance(token, userAddress);
    },
    enabled: enabled && !!client && !!token && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { balance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useShareBalance.ts
var import_react_query19 = require("@tanstack/react-query");
var import_wagmi5 = require("wagmi");
function useShareBalance(vault, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi5.useAccount)();
  const userAddress = account ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query19.useQuery)({
    queryKey: ["yo-share-balance", vaultAddress, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getShareBalance(vaultAddress, userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { shares: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useAllowance.ts
var import_react_query20 = require("@tanstack/react-query");
var import_wagmi6 = require("wagmi");
function useAllowance(token, spender, owner, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi6.useAccount)();
  const ownerAddress = owner ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query20.useQuery)({
    queryKey: ["yo-allowance", token, ownerAddress, spender, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!token || !ownerAddress || !spender) throw new Error("Missing parameters");
      return client.getAllowance(token, ownerAddress, spender);
    },
    enabled: enabled && !!client && !!token && !!ownerAddress && !!spender,
    staleTime: 15e3
  });
  return { allowance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserHistory.ts
var import_react_query21 = require("@tanstack/react-query");
var import_wagmi7 = require("wagmi");
function useUserHistory(vault, user, options) {
  const { limit, enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi7.useAccount)();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query21.useQuery)({
    queryKey: ["yo-user-history", vaultAddress, account, limit, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not available");
      return client.getUserHistory(vaultAddress, account, limit);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4
  });
  return { history: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPerformance.ts
var import_react_query22 = require("@tanstack/react-query");
var import_wagmi8 = require("wagmi");
function useUserPerformance(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi8.useAccount)();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query22.useQuery)({
    queryKey: ["yo-user-performance", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserPerformance(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { performance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserSnapshots.ts
var import_react_query23 = require("@tanstack/react-query");
var import_wagmi9 = require("wagmi");
function useUserSnapshots(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi9.useAccount)();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query23.useQuery)({
    queryKey: ["yo-user-snapshots", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserSnapshots(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 6e4
  });
  return { snapshots: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserBalances.ts
var import_react_query24 = require("@tanstack/react-query");
var import_wagmi10 = require("wagmi");
function useUserBalances(user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi10.useAccount)();
  const account = user ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query24.useQuery)({
    queryKey: ["yo-user-balances", account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserBalances(account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 6e4,
    refetchInterval: 6e4
  });
  return { balances: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePendingRedemptions.ts
var import_react_query25 = require("@tanstack/react-query");
var import_wagmi11 = require("wagmi");
function usePendingRedemptions(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi11.useAccount)();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = (0, import_react_query25.useQuery)({
    queryKey: ["yo-pending-redemptions", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getPendingRedemptions(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4
  });
  return { pendingRedemptions: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserRewards.ts
var import_react_query26 = require("@tanstack/react-query");
var import_wagmi12 = require("wagmi");
function useUserRewards(tokenAddress, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi12.useAccount)();
  const account = user ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query26.useQuery)({
    queryKey: ["yo-user-rewards", account, tokenAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserRewardsByAsset(account, tokenAddress);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 2 * 6e4
  });
  return { rewards: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useLeaderboard.ts
var import_react_query27 = require("@tanstack/react-query");
function useLeaderboard(period, tokenAddress, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = (0, import_react_query27.useQuery)({
    queryKey: ["yo-leaderboard", period, tokenAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return period === "weekly" ? client.getWeeklyRewards(tokenAddress) : client.getAllTimeRewards(tokenAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useMerklCampaigns.ts
var import_react_query28 = require("@tanstack/react-query");
function useMerklCampaigns(options) {
  const { status, enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = (0, import_react_query28.useQuery)({
    queryKey: ["merkl-campaigns", status, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getMerklCampaigns({ status });
    },
    enabled: enabled && !!client,
    staleTime: 6e4,
    refetchInterval: 12e4
  });
  return { campaigns: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useMerklRewards.ts
var import_react_query29 = require("@tanstack/react-query");
var import_wagmi13 = require("wagmi");
function useMerklRewards(account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = (0, import_wagmi13.useAccount)();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = (0, import_react_query29.useQuery)({
    queryKey: ["merkl-rewards", userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not available");
      return client.getClaimableRewards(userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 6e4,
    refetchInterval: 12e4
  });
  const totalClaimable = data ? client?.getMerklTotalClaimable(data) ?? 0n : 0n;
  const hasClaimable = data ? client?.hasMerklClaimableRewards(data) ?? false : false;
  return {
    rewards: data ?? void 0,
    totalClaimable,
    hasClaimable,
    isLoading,
    isError,
    error: error ?? null,
    refetch
  };
}

// src/hooks/useDeposit.ts
var import_react_query30 = require("@tanstack/react-query");
var import_react2 = require("react");
var import_wagmi14 = require("wagmi");
function useDeposit(options) {
  const { defaultSlippageBps, onError: globalOnError } = useYieldConfig();
  const { vault, slippageBps = defaultSlippageBps, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account, chainId: walletChainId } = (0, import_wagmi14.useAccount)();
  const queryClient = (0, import_react_query30.useQueryClient)();
  const vaultAddress = resolveVaultAddress(vault);
  const { switchChainAsync } = (0, import_wagmi14.useSwitchChain)();
  const { sendTransactionAsync } = (0, import_wagmi14.useSendTransaction)();
  const [hash, setHash] = (0, import_react2.useState)();
  const [approveHash, setApproveHash] = (0, import_react2.useState)();
  const [step, setStep] = (0, import_react2.useState)("idle");
  const [error, setError] = (0, import_react2.useState)(null);
  const { isSuccess } = (0, import_wagmi14.useWaitForTransactionReceipt)({ hash, query: { enabled: !!hash } });
  (0, import_react2.useEffect)(() => {
    if (isSuccess && hash) {
      setStep("success");
      queryClient.invalidateQueries({ queryKey: ["yo-user-position", vaultAddress, account] });
      queryClient.invalidateQueries({ queryKey: ["yo-vault-state", vaultAddress] });
      queryClient.invalidateQueries({ queryKey: ["yo-token-balance"] });
      queryClient.invalidateQueries({ queryKey: ["yo-share-balance"] });
      onConfirmed?.(hash);
    }
  }, [isSuccess, hash, queryClient, vaultAddress, account, onConfirmed]);
  const deposit = (0, import_react2.useCallback)(
    async (params) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setStep("idle");
      setError(null);
      try {
        if (params.chainId && walletChainId !== params.chainId) {
          setStep("switching-chain");
          await switchChainAsync({ chainId: params.chainId });
        }
        setStep("approving");
        const txs = await client.prepareDepositWithApproval({
          vault: vaultAddress,
          token: params.token,
          owner: account,
          recipient: account,
          amount: params.amount,
          slippageBps,
          chainId: params.chainId
        });
        if (txs.length > 1) {
          const approveTx = txs[0];
          const appHash = await sendTransactionAsync({ to: approveTx.to, data: approveTx.data, value: approveTx.value });
          setApproveHash(appHash);
          await client.waitForTransaction(appHash);
        }
        setStep("depositing");
        const depositTx = txs[txs.length - 1];
        const depositHash = await sendTransactionAsync({ to: depositTx.to, data: depositTx.data, value: depositTx.value });
        setHash(depositHash);
        setStep("waiting");
        onSubmitted?.(depositHash);
        return depositHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Deposit failed");
        setStep("error");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      }
    },
    [client, account, walletChainId, vaultAddress, slippageBps, switchChainAsync, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = (0, import_react2.useCallback)(() => {
    setHash(void 0);
    setApproveHash(void 0);
    setError(null);
    setStep("idle");
  }, []);
  return {
    deposit,
    step,
    isLoading: step !== "idle" && step !== "success" && step !== "error",
    isError: step === "error",
    error,
    isSuccess,
    hash,
    approveHash,
    reset
  };
}

// src/hooks/useRedeem.ts
var import_react_query31 = require("@tanstack/react-query");
var import_react3 = require("react");
var import_wagmi15 = require("wagmi");
function useRedeem(options) {
  const { onError: globalOnError } = useYieldConfig();
  const { vault, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account } = (0, import_wagmi15.useAccount)();
  const queryClient = (0, import_react_query31.useQueryClient)();
  const vaultAddress = resolveVaultAddress(vault);
  const { sendTransactionAsync } = (0, import_wagmi15.useSendTransaction)();
  const [hash, setHash] = (0, import_react3.useState)();
  const [approveHash, setApproveHash] = (0, import_react3.useState)();
  const [step, setStep] = (0, import_react3.useState)("idle");
  const [error, setError] = (0, import_react3.useState)(null);
  const [receipt, setReceipt] = (0, import_react3.useState)();
  const { isSuccess } = (0, import_wagmi15.useWaitForTransactionReceipt)({ hash, query: { enabled: !!hash } });
  (0, import_react3.useEffect)(() => {
    if (!isSuccess || !hash || !client) return;
    setStep("success");
    queryClient.invalidateQueries({ queryKey: ["yo-user-position", vaultAddress, account] });
    queryClient.invalidateQueries({ queryKey: ["yo-vault-state", vaultAddress] });
    queryClient.invalidateQueries({ queryKey: ["yo-share-balance"] });
    queryClient.invalidateQueries({ queryKey: ["yo-pending-redemptions"] });
    onConfirmed?.(hash);
    client.waitForRedeemReceipt(hash).then(setReceipt).catch(() => {
    });
  }, [isSuccess, hash, client, queryClient, vaultAddress, account, onConfirmed]);
  const redeem = (0, import_react3.useCallback)(
    async (shares) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setStep("idle");
      setError(null);
      try {
        setStep("approving");
        const txs = await client.prepareRedeemWithApproval({
          vault: vaultAddress,
          shares,
          owner: account,
          recipient: account
        });
        if (txs.length > 1) {
          const approveTx = txs[0];
          const appHash = await sendTransactionAsync({ to: approveTx.to, data: approveTx.data, value: approveTx.value });
          setApproveHash(appHash);
          await client.waitForTransaction(appHash);
        }
        setStep("redeeming");
        const redeemTx = txs[txs.length - 1];
        const redeemHash = await sendTransactionAsync({ to: redeemTx.to, data: redeemTx.data, value: redeemTx.value });
        setHash(redeemHash);
        setStep("waiting");
        onSubmitted?.(redeemHash);
        return redeemHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Redeem failed");
        setStep("error");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      }
    },
    [client, account, vaultAddress, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = (0, import_react3.useCallback)(() => {
    setHash(void 0);
    setApproveHash(void 0);
    setError(null);
    setStep("idle");
    setReceipt(void 0);
  }, []);
  return {
    redeem,
    step,
    isLoading: step !== "idle" && step !== "success" && step !== "error",
    isError: step === "error",
    error,
    isSuccess,
    hash,
    approveHash,
    instant: receipt?.instant,
    assetsOrRequestId: receipt?.assetsOrRequestId,
    reset
  };
}

// src/hooks/useApprove.ts
var import_core3 = require("@yo-protocol/core");
var import_react4 = require("react");
var import_viem = require("viem");
var import_wagmi16 = require("wagmi");
function useApprove(options) {
  const { token, spender = import_core3.YO_GATEWAY_ADDRESS, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account } = (0, import_wagmi16.useAccount)();
  const { sendTransactionAsync } = (0, import_wagmi16.useSendTransaction)();
  const [hash, setHash] = (0, import_react4.useState)();
  const [isLoading, setIsLoading] = (0, import_react4.useState)(false);
  const [error, setError] = (0, import_react4.useState)(null);
  const { isSuccess } = (0, import_wagmi16.useWaitForTransactionReceipt)({ hash, query: { enabled: !!hash } });
  (0, import_react4.useEffect)(() => {
    if (isSuccess && hash) onConfirmed?.(hash);
  }, [isSuccess, hash, onConfirmed]);
  const approve = (0, import_react4.useCallback)(
    async (amount) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setIsLoading(true);
      setError(null);
      try {
        const tx = client.prepareApprove({ token, spender, amount });
        const txHash = await sendTransactionAsync({ to: tx.to, data: tx.data, value: tx.value });
        setHash(txHash);
        onSubmitted?.(txHash);
        return txHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Approval failed");
        setError(e);
        onError?.(e);
        throw e;
      } finally {
        setIsLoading(false);
      }
    },
    [client, account, token, spender, sendTransactionAsync, onSubmitted, onError]
  );
  const approveMax = (0, import_react4.useCallback)(() => approve(import_viem.maxUint256), [approve]);
  const reset = (0, import_react4.useCallback)(() => {
    setHash(void 0);
    setError(null);
    setIsLoading(false);
  }, []);
  return { approve, approveMax, isLoading, isError: !!error, error, isSuccess, hash, reset };
}

// src/hooks/useClaimMerklRewards.ts
var import_react_query32 = require("@tanstack/react-query");
var import_react5 = require("react");
var import_wagmi17 = require("wagmi");
function useClaimMerklRewards(options) {
  const { onError: globalOnError } = useYieldConfig();
  const { onSubmitted, onConfirmed, onError } = options ?? {};
  const client = useYoClient();
  const { address: account } = (0, import_wagmi17.useAccount)();
  const queryClient = (0, import_react_query32.useQueryClient)();
  const { sendTransactionAsync } = (0, import_wagmi17.useSendTransaction)();
  const [hash, setHash] = (0, import_react5.useState)();
  const [isLoading, setIsLoading] = (0, import_react5.useState)(false);
  const [error, setError] = (0, import_react5.useState)(null);
  const { isSuccess } = (0, import_wagmi17.useWaitForTransactionReceipt)({ hash, query: { enabled: !!hash } });
  (0, import_react5.useEffect)(() => {
    if (isSuccess && hash) {
      queryClient.invalidateQueries({ queryKey: ["merkl-rewards", account] });
      onConfirmed?.(hash);
    }
  }, [isSuccess, hash, queryClient, account, onConfirmed]);
  const claim = (0, import_react5.useCallback)(
    async (chainRewards) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setIsLoading(true);
      setError(null);
      try {
        const tx = client.prepareClaimMerklRewards(account, chainRewards);
        const txHash = await sendTransactionAsync({ to: tx.to, data: tx.data, value: tx.value });
        setHash(txHash);
        onSubmitted?.(txHash);
        return txHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Claim failed");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      } finally {
        setIsLoading(false);
      }
    },
    [client, account, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = (0, import_react5.useCallback)(() => {
    setHash(void 0);
    setError(null);
    setIsLoading(false);
  }, []);
  return { claim, isLoading, isError: !!error, error, isSuccess, hash, reset };
}

// src/hooks/useCreateYoClient.ts
var import_core4 = require("@yo-protocol/core");
var import_react6 = require("react");
function useCreateYoClient(options) {
  const { chainId, partnerId, publicClients } = options;
  return (0, import_react6.useMemo)(() => {
    if (!(0, import_core4.isSupportedChain)(chainId)) return null;
    return (0, import_core4.createYoClient)({ chainId, partnerId, publicClients });
  }, [chainId, partnerId, publicClients]);
}
