// @yo-protocol/react/dist/index.js
"use client";
"use client";

// src/context.tsx
import { createYoClient, isSupportedChain } from "@yo-protocol/core";
import { createContext, useContext, useMemo } from "react";
import { useChainId, usePublicClient } from "wagmi";
import { jsx } from "react/jsx-runtime";
var YieldContext = createContext({
  ready: false,
  defaultSlippageBps: 50
});
function YieldProvider({
  children,
  partnerId,
  defaultSlippageBps = 50,
  onError
}) {
  const value = useMemo(
    () => ({ ready: true, partnerId, defaultSlippageBps, onError }),
    [partnerId, defaultSlippageBps, onError]
  );
  return /* @__PURE__ */ jsx(YieldContext.Provider, { value, children });
}
function useYieldConfig() {
  const ctx = useContext(YieldContext);
  if (!ctx.ready) {
    throw new Error("useYieldConfig must be used within a YieldProvider");
  }
  return ctx;
}
function useYoClient() {
  const { ready, partnerId } = useContext(YieldContext);
  const chainId = useChainId();
  const publicClient = usePublicClient();
  return useMemo(() => {
    if (!ready || !isSupportedChain(chainId) || !publicClient) {
      return null;
    }
    return createYoClient({
      chainId,
      partnerId,
      publicClients: {
        [chainId]: publicClient
      }
    });
  }, [ready, chainId, publicClient, partnerId]);
}

// src/hooks/useVaultState.ts
import { useQuery } from "@tanstack/react-query";

// src/utils/vault.ts
import { VAULTS } from "@yo-protocol/core";
function isVaultId(value) {
  return value in VAULTS;
}
function resolveVaultAddress(vault) {
  if (isVaultId(vault)) {
    return VAULTS[vault].address;
  }
  return vault;
}

// src/hooks/useVaultState.ts
function useVaultState(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery({
    queryKey: ["yo-vault-state", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultState(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { vaultState: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePreviewDeposit.ts
import { useQuery as useQuery2 } from "@tanstack/react-query";
function usePreviewDeposit(vault, assets, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery2({
    queryKey: ["yo-preview-deposit", vaultAddress, assets?.toString(), client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.previewDeposit(vaultAddress, assets);
    },
    enabled: enabled && !!client && assets !== void 0 && assets > 0n,
    staleTime: 15e3
  });
  return { shares: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePreviewRedeem.ts
import { useQuery as useQuery3 } from "@tanstack/react-query";
function usePreviewRedeem(vault, shares, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery3({
    queryKey: ["yo-preview-redeem", vaultAddress, shares?.toString(), client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.previewRedeem(vaultAddress, shares);
    },
    enabled: enabled && !!client && shares !== void 0 && shares > 0n,
    staleTime: 15e3
  });
  return { assets: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaults.ts
import { useQuery as useQuery4 } from "@tanstack/react-query";
function useVaults(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error } = useQuery4({
    queryKey: ["yo-vaults"],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaults({ secondary: false });
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { vaults: data ?? [], isLoading, isError, error: error ?? null };
}

// src/hooks/useVaultHistory.ts
import { useQuery as useQuery5 } from "@tanstack/react-query";
function useVaultHistory(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const yieldQuery = useQuery5({
    queryKey: ["yo-vault-yield-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultYieldHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  const tvlQuery = useQuery5({
    queryKey: ["yo-vault-tvl-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultTvlHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return {
    yieldHistory: yieldQuery.data ?? [],
    tvlHistory: tvlQuery.data ?? [],
    isLoading: yieldQuery.isLoading || tvlQuery.isLoading,
    isError: yieldQuery.isError || tvlQuery.isError,
    error: yieldQuery.error ?? tvlQuery.error ?? null,
    refetch: () => {
      yieldQuery.refetch();
      tvlQuery.refetch();
    }
  };
}

// src/hooks/useVaultTransactionHistory.ts
import { useQuery as useQuery6 } from "@tanstack/react-query";
function useVaultTransactionHistory(vault, options) {
  const { limit, cursor, filters, allNetworks = false, enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery6({
    queryKey: ["yo-vault-tx-history", vaultAddress, allNetworks, cursor, limit, filters, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (allNetworks) {
        return client.getVaultHistoryAllNetworks(vaultAddress, { cursor, limit, filters });
      }
      return client.getVaultHistory(vaultAddress, limit);
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { history: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useGlobalVaultHistory.ts
import { useQuery as useQuery7 } from "@tanstack/react-query";
function useGlobalVaultHistory(options) {
  const { cursor, limit, filters, enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = useQuery7({
    queryKey: ["yo-global-vault-history", cursor, limit, filters, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getGlobalVaultHistory({ cursor, limit, filters });
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { history: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPerformance.ts
import { useQuery as useQuery8 } from "@tanstack/react-query";
function useVaultPerformance(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery8({
    queryKey: ["yo-vault-performance", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPerformance(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { performance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPercentile.ts
import { useQuery as useQuery9 } from "@tanstack/react-query";
function useVaultPercentile(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery9({
    queryKey: ["yo-vault-percentile", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPercentile(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { percentile: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePerformanceBenchmark.ts
import { useQuery as useQuery10 } from "@tanstack/react-query";
function usePerformanceBenchmark(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery10({
    queryKey: ["yo-performance-benchmark", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getPerformanceBenchmark(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { benchmark: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultAllocations.ts
import { useQuery as useQuery11 } from "@tanstack/react-query";
function useVaultAllocations(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery11({
    queryKey: ["yo-vault-allocations", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultAllocationsTimeSeries(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { allocations: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useVaultPendingRedeems.ts
import { useQuery as useQuery12 } from "@tanstack/react-query";
function useVaultPendingRedeems(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery12({
    queryKey: ["yo-vault-pending-redeems", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getVaultPendingRedeems(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 3e4
  });
  return { pendingRedeems: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useSharePriceHistory.ts
import { useQuery as useQuery13 } from "@tanstack/react-query";
function useSharePriceHistory(vault, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery13({
    queryKey: ["yo-share-price-history", vaultAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getSharePriceHistory(vaultAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { history: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useTotalTvl.ts
import { useQuery as useQuery14 } from "@tanstack/react-query";
function useTotalTvl(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = useQuery14({
    queryKey: ["yo-total-tvl", client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getTotalTvlTimeseries();
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { tvl: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePrices.ts
import { useQuery as useQuery15 } from "@tanstack/react-query";
function usePrices(options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = useQuery15({
    queryKey: ["yo-prices", client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getPrices();
    },
    enabled: enabled && !!client,
    staleTime: 6e4
  });
  return { prices: data ?? {}, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPosition.ts
import { useQuery as useQuery16 } from "@tanstack/react-query";
import { useAccount } from "wagmi";
function useUserPosition(vault, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount();
  const userAddress = account ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery16({
    queryKey: ["yo-user-position", vaultAddress, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getUserPosition(vaultAddress, userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { position: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPositions.ts
import { useQuery as useQuery17 } from "@tanstack/react-query";
import { useAccount as useAccount2 } from "wagmi";
function useUserPositions(account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount2();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery17({
    queryKey: ["yo-user-positions", userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getUserPositionsAllChains(userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { positions: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useTokenBalance.ts
import { useQuery as useQuery18 } from "@tanstack/react-query";
import { useAccount as useAccount3 } from "wagmi";
function useTokenBalance(token, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount3();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery18({
    queryKey: ["yo-token-balance", token, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!token || !userAddress) throw new Error("Missing parameters");
      return client.getTokenBalance(token, userAddress);
    },
    enabled: enabled && !!client && !!token && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { balance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useShareBalance.ts
import { useQuery as useQuery19 } from "@tanstack/react-query";
import { useAccount as useAccount4 } from "wagmi";
function useShareBalance(vault, account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount4();
  const userAddress = account ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery19({
    queryKey: ["yo-share-balance", vaultAddress, userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not connected");
      return client.getShareBalance(vaultAddress, userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 15e3,
    refetchInterval: 3e4
  });
  return { shares: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useAllowance.ts
import { useQuery as useQuery20 } from "@tanstack/react-query";
import { useAccount as useAccount5 } from "wagmi";
function useAllowance(token, spender, owner, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount5();
  const ownerAddress = owner ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery20({
    queryKey: ["yo-allowance", token, ownerAddress, spender, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!token || !ownerAddress || !spender) throw new Error("Missing parameters");
      return client.getAllowance(token, ownerAddress, spender);
    },
    enabled: enabled && !!client && !!token && !!ownerAddress && !!spender,
    staleTime: 15e3
  });
  return { allowance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserHistory.ts
import { useQuery as useQuery21 } from "@tanstack/react-query";
import { useAccount as useAccount6 } from "wagmi";
function useUserHistory(vault, user, options) {
  const { limit, enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount6();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery21({
    queryKey: ["yo-user-history", vaultAddress, account, limit, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not available");
      return client.getUserHistory(vaultAddress, account, limit);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4
  });
  return { history: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserPerformance.ts
import { useQuery as useQuery22 } from "@tanstack/react-query";
import { useAccount as useAccount7 } from "wagmi";
function useUserPerformance(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount7();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery22({
    queryKey: ["yo-user-performance", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserPerformance(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4,
    refetchInterval: 6e4
  });
  return { performance: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserSnapshots.ts
import { useQuery as useQuery23 } from "@tanstack/react-query";
import { useAccount as useAccount8 } from "wagmi";
function useUserSnapshots(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount8();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery23({
    queryKey: ["yo-user-snapshots", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserSnapshots(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 6e4
  });
  return { snapshots: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserBalances.ts
import { useQuery as useQuery24 } from "@tanstack/react-query";
import { useAccount as useAccount9 } from "wagmi";
function useUserBalances(user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount9();
  const account = user ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery24({
    queryKey: ["yo-user-balances", account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserBalances(account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 6e4,
    refetchInterval: 6e4
  });
  return { balances: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/usePendingRedemptions.ts
import { useQuery as useQuery25 } from "@tanstack/react-query";
import { useAccount as useAccount10 } from "wagmi";
function usePendingRedemptions(vault, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount10();
  const account = user ?? connectedAddress;
  const vaultAddress = resolveVaultAddress(vault);
  const { data, isLoading, isError, error, refetch } = useQuery25({
    queryKey: ["yo-pending-redemptions", vaultAddress, account, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getPendingRedemptions(vaultAddress, account);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 3e4
  });
  return { pendingRedemptions: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useUserRewards.ts
import { useQuery as useQuery26 } from "@tanstack/react-query";
import { useAccount as useAccount11 } from "wagmi";
function useUserRewards(tokenAddress, user, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount11();
  const account = user ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery26({
    queryKey: ["yo-user-rewards", account, tokenAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      return client.getUserRewardsByAsset(account, tokenAddress);
    },
    enabled: enabled && !!client && !!account,
    staleTime: 2 * 6e4
  });
  return { rewards: data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useLeaderboard.ts
import { useQuery as useQuery27 } from "@tanstack/react-query";
function useLeaderboard(period, tokenAddress, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = useQuery27({
    queryKey: ["yo-leaderboard", period, tokenAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return period === "weekly" ? client.getWeeklyRewards(tokenAddress) : client.getAllTimeRewards(tokenAddress);
    },
    enabled: enabled && !!client,
    staleTime: 5 * 6e4
  });
  return { data, isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useMerklCampaigns.ts
import { useQuery as useQuery28 } from "@tanstack/react-query";
function useMerklCampaigns(options) {
  const { status, enabled = true } = options ?? {};
  const client = useYoClient();
  const { data, isLoading, isError, error, refetch } = useQuery28({
    queryKey: ["merkl-campaigns", status, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      return client.getMerklCampaigns({ status });
    },
    enabled: enabled && !!client,
    staleTime: 6e4,
    refetchInterval: 12e4
  });
  return { campaigns: data ?? [], isLoading, isError, error: error ?? null, refetch };
}

// src/hooks/useMerklRewards.ts
import { useQuery as useQuery29 } from "@tanstack/react-query";
import { useAccount as useAccount12 } from "wagmi";
function useMerklRewards(account, options) {
  const { enabled = true } = options ?? {};
  const client = useYoClient();
  const { address: connectedAddress } = useAccount12();
  const userAddress = account ?? connectedAddress;
  const { data, isLoading, isError, error, refetch } = useQuery29({
    queryKey: ["merkl-rewards", userAddress, client?.chainId],
    queryFn: () => {
      if (!client) throw new Error("Client not available");
      if (!userAddress) throw new Error("Account not available");
      return client.getClaimableRewards(userAddress);
    },
    enabled: enabled && !!client && !!userAddress,
    staleTime: 6e4,
    refetchInterval: 12e4
  });
  const totalClaimable = data ? client?.getMerklTotalClaimable(data) ?? 0n : 0n;
  const hasClaimable = data ? client?.hasMerklClaimableRewards(data) ?? false : false;
  return {
    rewards: data ?? void 0,
    totalClaimable,
    hasClaimable,
    isLoading,
    isError,
    error: error ?? null,
    refetch
  };
}

// src/hooks/useDeposit.ts
import { useQueryClient } from "@tanstack/react-query";
import { useCallback, useEffect, useState } from "react";
import { useAccount as useAccount13, useSendTransaction, useSwitchChain, useWaitForTransactionReceipt } from "wagmi";
function useDeposit(options) {
  const { defaultSlippageBps, onError: globalOnError } = useYieldConfig();
  const { vault, slippageBps = defaultSlippageBps, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account, chainId: walletChainId } = useAccount13();
  const queryClient = useQueryClient();
  const vaultAddress = resolveVaultAddress(vault);
  const { switchChainAsync } = useSwitchChain();
  const { sendTransactionAsync } = useSendTransaction();
  const [hash, setHash] = useState();
  const [approveHash, setApproveHash] = useState();
  const [step, setStep] = useState("idle");
  const [error, setError] = useState(null);
  const { isSuccess } = useWaitForTransactionReceipt({ hash, query: { enabled: !!hash } });
  useEffect(() => {
    if (isSuccess && hash) {
      setStep("success");
      queryClient.invalidateQueries({ queryKey: ["yo-user-position", vaultAddress, account] });
      queryClient.invalidateQueries({ queryKey: ["yo-vault-state", vaultAddress] });
      queryClient.invalidateQueries({ queryKey: ["yo-token-balance"] });
      queryClient.invalidateQueries({ queryKey: ["yo-share-balance"] });
      onConfirmed?.(hash);
    }
  }, [isSuccess, hash, queryClient, vaultAddress, account, onConfirmed]);
  const deposit = useCallback(
    async (params) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setStep("idle");
      setError(null);
      try {
        if (params.chainId && walletChainId !== params.chainId) {
          setStep("switching-chain");
          await switchChainAsync({ chainId: params.chainId });
        }
        setStep("approving");
        const txs = await client.prepareDepositWithApproval({
          vault: vaultAddress,
          token: params.token,
          owner: account,
          recipient: account,
          amount: params.amount,
          slippageBps,
          chainId: params.chainId
        });
        if (txs.length > 1) {
          const approveTx = txs[0];
          const appHash = await sendTransactionAsync({ to: approveTx.to, data: approveTx.data, value: approveTx.value });
          setApproveHash(appHash);
          await client.waitForTransaction(appHash);
        }
        setStep("depositing");
        const depositTx = txs[txs.length - 1];
        const depositHash = await sendTransactionAsync({ to: depositTx.to, data: depositTx.data, value: depositTx.value });
        setHash(depositHash);
        setStep("waiting");
        onSubmitted?.(depositHash);
        return depositHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Deposit failed");
        setStep("error");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      }
    },
    [client, account, walletChainId, vaultAddress, slippageBps, switchChainAsync, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = useCallback(() => {
    setHash(void 0);
    setApproveHash(void 0);
    setError(null);
    setStep("idle");
  }, []);
  return {
    deposit,
    step,
    isLoading: step !== "idle" && step !== "success" && step !== "error",
    isError: step === "error",
    error,
    isSuccess,
    hash,
    approveHash,
    reset
  };
}

// src/hooks/useRedeem.ts
import { useQueryClient as useQueryClient2 } from "@tanstack/react-query";
import { useCallback as useCallback2, useEffect as useEffect2, useState as useState2 } from "react";
import { useAccount as useAccount14, useSendTransaction as useSendTransaction2, useWaitForTransactionReceipt as useWaitForTransactionReceipt2 } from "wagmi";
function useRedeem(options) {
  const { onError: globalOnError } = useYieldConfig();
  const { vault, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account } = useAccount14();
  const queryClient = useQueryClient2();
  const vaultAddress = resolveVaultAddress(vault);
  const { sendTransactionAsync } = useSendTransaction2();
  const [hash, setHash] = useState2();
  const [approveHash, setApproveHash] = useState2();
  const [step, setStep] = useState2("idle");
  const [error, setError] = useState2(null);
  const [receipt, setReceipt] = useState2();
  const { isSuccess } = useWaitForTransactionReceipt2({ hash, query: { enabled: !!hash } });
  useEffect2(() => {
    if (!isSuccess || !hash || !client) return;
    setStep("success");
    queryClient.invalidateQueries({ queryKey: ["yo-user-position", vaultAddress, account] });
    queryClient.invalidateQueries({ queryKey: ["yo-vault-state", vaultAddress] });
    queryClient.invalidateQueries({ queryKey: ["yo-share-balance"] });
    queryClient.invalidateQueries({ queryKey: ["yo-pending-redemptions"] });
    onConfirmed?.(hash);
    client.waitForRedeemReceipt(hash).then(setReceipt).catch(() => {
    });
  }, [isSuccess, hash, client, queryClient, vaultAddress, account, onConfirmed]);
  const redeem = useCallback2(
    async (shares) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setStep("idle");
      setError(null);
      try {
        setStep("approving");
        const txs = await client.prepareRedeemWithApproval({
          vault: vaultAddress,
          shares,
          owner: account,
          recipient: account
        });
        if (txs.length > 1) {
          const approveTx = txs[0];
          const appHash = await sendTransactionAsync({ to: approveTx.to, data: approveTx.data, value: approveTx.value });
          setApproveHash(appHash);
          await client.waitForTransaction(appHash);
        }
        setStep("redeeming");
        const redeemTx = txs[txs.length - 1];
        const redeemHash = await sendTransactionAsync({ to: redeemTx.to, data: redeemTx.data, value: redeemTx.value });
        setHash(redeemHash);
        setStep("waiting");
        onSubmitted?.(redeemHash);
        return redeemHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Redeem failed");
        setStep("error");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      }
    },
    [client, account, vaultAddress, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = useCallback2(() => {
    setHash(void 0);
    setApproveHash(void 0);
    setError(null);
    setStep("idle");
    setReceipt(void 0);
  }, []);
  return {
    redeem,
    step,
    isLoading: step !== "idle" && step !== "success" && step !== "error",
    isError: step === "error",
    error,
    isSuccess,
    hash,
    approveHash,
    instant: receipt?.instant,
    assetsOrRequestId: receipt?.assetsOrRequestId,
    reset
  };
}

// src/hooks/useApprove.ts
import { YO_GATEWAY_ADDRESS } from "@yo-protocol/core";
import { useCallback as useCallback3, useEffect as useEffect3, useState as useState3 } from "react";
import { maxUint256 } from "viem";
import { useAccount as useAccount15, useSendTransaction as useSendTransaction3, useWaitForTransactionReceipt as useWaitForTransactionReceipt3 } from "wagmi";
function useApprove(options) {
  const { token, spender = YO_GATEWAY_ADDRESS, onSubmitted, onConfirmed, onError } = options;
  const client = useYoClient();
  const { address: account } = useAccount15();
  const { sendTransactionAsync } = useSendTransaction3();
  const [hash, setHash] = useState3();
  const [isLoading, setIsLoading] = useState3(false);
  const [error, setError] = useState3(null);
  const { isSuccess } = useWaitForTransactionReceipt3({ hash, query: { enabled: !!hash } });
  useEffect3(() => {
    if (isSuccess && hash) onConfirmed?.(hash);
  }, [isSuccess, hash, onConfirmed]);
  const approve = useCallback3(
    async (amount) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setIsLoading(true);
      setError(null);
      try {
        const tx = client.prepareApprove({ token, spender, amount });
        const txHash = await sendTransactionAsync({ to: tx.to, data: tx.data, value: tx.value });
        setHash(txHash);
        onSubmitted?.(txHash);
        return txHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Approval failed");
        setError(e);
        onError?.(e);
        throw e;
      } finally {
        setIsLoading(false);
      }
    },
    [client, account, token, spender, sendTransactionAsync, onSubmitted, onError]
  );
  const approveMax = useCallback3(() => approve(maxUint256), [approve]);
  const reset = useCallback3(() => {
    setHash(void 0);
    setError(null);
    setIsLoading(false);
  }, []);
  return { approve, approveMax, isLoading, isError: !!error, error, isSuccess, hash, reset };
}

// src/hooks/useClaimMerklRewards.ts
import { useQueryClient as useQueryClient3 } from "@tanstack/react-query";
import { useCallback as useCallback4, useEffect as useEffect4, useState as useState4 } from "react";
import { useAccount as useAccount16, useSendTransaction as useSendTransaction4, useWaitForTransactionReceipt as useWaitForTransactionReceipt4 } from "wagmi";
function useClaimMerklRewards(options) {
  const { onError: globalOnError } = useYieldConfig();
  const { onSubmitted, onConfirmed, onError } = options ?? {};
  const client = useYoClient();
  const { address: account } = useAccount16();
  const queryClient = useQueryClient3();
  const { sendTransactionAsync } = useSendTransaction4();
  const [hash, setHash] = useState4();
  const [isLoading, setIsLoading] = useState4(false);
  const [error, setError] = useState4(null);
  const { isSuccess } = useWaitForTransactionReceipt4({ hash, query: { enabled: !!hash } });
  useEffect4(() => {
    if (isSuccess && hash) {
      queryClient.invalidateQueries({ queryKey: ["merkl-rewards", account] });
      onConfirmed?.(hash);
    }
  }, [isSuccess, hash, queryClient, account, onConfirmed]);
  const claim = useCallback4(
    async (chainRewards) => {
      if (!client) throw new Error("Client not available");
      if (!account) throw new Error("Account not connected");
      setIsLoading(true);
      setError(null);
      try {
        const tx = client.prepareClaimMerklRewards(account, chainRewards);
        const txHash = await sendTransactionAsync({ to: tx.to, data: tx.data, value: tx.value });
        setHash(txHash);
        onSubmitted?.(txHash);
        return txHash;
      } catch (err) {
        const e = err instanceof Error ? err : new Error("Claim failed");
        setError(e);
        onError?.(e);
        globalOnError?.(e);
        throw e;
      } finally {
        setIsLoading(false);
      }
    },
    [client, account, sendTransactionAsync, onSubmitted, onError, globalOnError]
  );
  const reset = useCallback4(() => {
    setHash(void 0);
    setError(null);
    setIsLoading(false);
  }, []);
  return { claim, isLoading, isError: !!error, error, isSuccess, hash, reset };
}

// src/hooks/useCreateYoClient.ts
import { createYoClient as createYoClient2, isSupportedChain as isSupportedChain2 } from "@yo-protocol/core";
import { useMemo as useMemo2 } from "react";
function useCreateYoClient(options) {
  const { chainId, partnerId, publicClients } = options;
  return useMemo2(() => {
    if (!isSupportedChain2(chainId)) return null;
    return createYoClient2({ chainId, partnerId, publicClients });
  }, [chainId, partnerId, publicClients]);
}
export {
  YieldProvider,
  useAllowance,
  useApprove,
  useClaimMerklRewards,
  useCreateYoClient,
  useDeposit,
  useGlobalVaultHistory,
  useLeaderboard,
  useMerklCampaigns,
  useMerklRewards,
  usePendingRedemptions,
  usePerformanceBenchmark,
  usePreviewDeposit,
  usePreviewRedeem,
  usePrices,
  useRedeem,
  useShareBalance,
  useSharePriceHistory,
  useTokenBalance,
  useTotalTvl,
  useUserBalances,
  useUserHistory,
  useUserPerformance,
  useUserPosition,
  useUserPositions,
  useUserRewards,
  useUserSnapshots,
  useVaultAllocations,
  useVaultHistory,
  useVaultPendingRedeems,
  useVaultPercentile,
  useVaultPerformance,
  useVaultState,
  useVaultTransactionHistory,
  useVaults,
  useYieldConfig,
  useYoClient
};
