// @yo-protocol/core/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  API_BASE_URL: () => API_BASE_URL,
  ApiClient: () => ApiClient,
  ApiError: () => ApiError,
  CHAIN_ID_TO_NETWORK: () => CHAIN_ID_TO_NETWORK,
  DEFAULT_ASSET_IDS: () => DEFAULT_ASSET_IDS,
  MERKL_API_BASE_URL: () => MERKL_API_BASE_URL,
  MERKL_CREATOR_ADDRESS: () => MERKL_CREATOR_ADDRESS,
  MERKL_DISTRIBUTOR_ADDRESSES: () => MERKL_DISTRIBUTOR_ADDRESSES,
  MerklApiClient: () => MerklApiClient,
  NETWORK_TO_CHAIN_ID: () => NETWORK_TO_CHAIN_ID,
  REDEEMER_ADDRESS: () => REDEEMER_ADDRESS,
  SUPPORTED_CHAINS: () => SUPPORTED_CHAINS,
  SUPPORTED_CHAIN_IDS: () => SUPPORTED_CHAIN_IDS,
  VAULTS: () => VAULTS,
  VAULT_REGISTRY_ADDRESS: () => VAULT_REGISTRY_ADDRESS,
  YO_GATEWAY_ADDRESS: () => YO_GATEWAY_ADDRESS,
  YO_ORACLE_ADDRESS: () => YO_ORACLE_ADDRESS,
  YO_TOKEN: () => YO_TOKEN,
  YoClient: () => YoClient,
  abs: () => abs,
  add: () => add,
  applySlippage: () => applySlippage,
  campaignHasPendingRewards: () => campaignHasPendingRewards,
  convertToAssets: () => convertToAssets,
  convertToShares: () => convertToShares,
  createApiClient: () => createApiClient,
  createMerklApiClient: () => createMerklApiClient,
  createYoClient: () => createYoClient,
  div: () => div,
  eq: () => eq,
  erc20Abi: () => erc20Abi,
  erc4626Abi: () => erc4626Abi,
  formatCompactNumber: () => formatCompactNumber,
  formatPercent: () => formatPercent,
  formatTokenAmount: () => formatTokenAmount,
  formatUsd: () => formatUsd,
  getAllTimeRewards: () => getAllTimeRewards,
  getAllVaults: () => getAllVaults,
  getAssetAllowance: () => getAssetAllowance,
  getChain: () => getChain,
  getClaimableAmount: () => getClaimableAmount,
  getClaimedAmount: () => getClaimedAmount,
  getClaimedAmounts: () => getClaimedAmounts,
  getGlobalVaultHistory: () => getGlobalVaultHistory,
  getIdleBalance: () => getIdleBalance,
  getPendingRewardSets: () => getPendingRewardSets,
  getPerformanceBenchmark: () => getPerformanceBenchmark,
  getPrices: () => getPrices,
  getShareAllowance: () => getShareAllowance,
  getShareBalance: () => getShareBalance,
  getSharePriceHistory: () => getSharePriceHistory,
  getTokenAllowance: () => getTokenAllowance,
  getTokenBalance: () => getTokenBalance,
  getTotalClaimable: () => getTotalClaimable,
  getTotalTvlTimeseries: () => getTotalTvlTimeseries,
  getUserBalances: () => getUserBalances,
  getUserPendingRedemptions: () => getUserPendingRedemptions,
  getUserPerformance: () => getUserPerformance,
  getUserPoints: () => getUserPoints,
  getUserPositionsAllChains: () => getUserPositionsAllChains,
  getUserRewardsByAsset: () => getUserRewardsByAsset,
  getUserSnapshots: () => getUserSnapshots,
  getUserTransactionHistory: () => getUserTransactionHistory,
  getUserVaultPosition: () => getUserVaultPosition,
  getVaultAllocationsTimeSeries: () => getVaultAllocationsTimeSeries,
  getVaultByAddress: () => getVaultByAddress,
  getVaultHistory: () => getVaultHistory,
  getVaultHistoryAllNetworks: () => getVaultHistoryAllNetworks,
  getVaultPendingRedeems: () => getVaultPendingRedeems,
  getVaultPercentile: () => getVaultPercentile,
  getVaultPerformance: () => getVaultPerformance,
  getVaultSnapshot: () => getVaultSnapshot,
  getVaultState: () => getVaultState,
  getVaultTvlTimeseries: () => getVaultTvlTimeseries,
  getVaultYieldTimeseries: () => getVaultYieldTimeseries,
  getVaults: () => getVaults,
  getVaultsForChain: () => getVaultsForChain,
  getWeeklyRewards: () => getWeeklyRewards,
  gt: () => gt,
  gte: () => gte,
  hasClaimableRewards: () => hasClaimableRewards,
  hasEnoughAllowance: () => hasEnoughAllowance,
  isNativeToken: () => isNativeToken,
  isPaused: () => isPaused,
  isSupportedChain: () => isSupportedChain,
  lt: () => lt,
  lte: () => lte,
  maxDeposit: () => maxDeposit,
  maxRedeem: () => maxRedeem,
  merklDistributorAbi: () => merklDistributorAbi,
  mul: () => mul,
  mulDiv: () => mulDiv,
  parseTokenAmount: () => parseTokenAmount,
  pausableAbi: () => pausableAbi,
  pow: () => pow,
  prepareApprove: () => prepareApprove,
  prepareClaimMerklRewards: () => prepareClaimMerklRewards,
  prepareClaimParams: () => prepareClaimParams,
  prepareDeposit: () => prepareDeposit,
  prepareDepositWithApproval: () => prepareDepositWithApproval,
  prepareRedeem: () => prepareRedeem,
  prepareRedeemWithApproval: () => prepareRedeemWithApproval,
  previewDeposit: () => previewDeposit,
  previewRedeem: () => previewRedeem,
  quoteConvertToAssets: () => quoteConvertToAssets,
  quoteConvertToShares: () => quoteConvertToShares,
  quotePreviewDeposit: () => quotePreviewDeposit,
  quotePreviewRedeem: () => quotePreviewRedeem,
  quotePreviewWithdraw: () => quotePreviewWithdraw,
  sqrt: () => sqrt,
  sub: () => sub,
  toBig: () => toBig,
  toBigInt: () => toBigInt,
  toFixed: () => toFixed,
  toNumber: () => toNumber,
  toStr: () => toStr,
  validateAddress: () => validateAddress,
  validateAmount: () => validateAmount,
  waitForRedeemReceipt: () => waitForRedeemReceipt,
  yoGatewayAbi: () => yoGatewayAbi
});
module.exports = __toCommonJS(index_exports);

// src/actions/merkl.ts
var import_viem = require("viem");
var import_chains2 = require("viem/chains");

// src/constants/merkl.ts
var import_chains = require("viem/chains");
var MERKL_API_BASE_URL = "https://api.merkl.xyz/v4";
var MERKL_CREATOR_ADDRESS = "0x8C9200d94Cf7A1B201068c4deDa6239F15FED480";
var MERKL_DISTRIBUTOR_ADDRESSES = {
  [import_chains.base.id]: "0x3Ef3D8bA38EBe18DB133cEc108f4D14CE00Dd9Ae"
};
var YO_TOKEN = {
  address: "0x1925450f5e5fB974b0AaE1F3408cF5286FBD1A72",
  symbol: "YO",
  name: "YO Token",
  decimals: 18,
  chainId: import_chains.base.id
};

// src/constants/merkl-abi.ts
var merklDistributorAbi = [
  {
    inputs: [
      { internalType: "address[]", name: "users", type: "address[]" },
      { internalType: "address[]", name: "tokens", type: "address[]" },
      { internalType: "uint256[]", name: "amounts", type: "uint256[]" },
      { internalType: "bytes32[][]", name: "proofs", type: "bytes32[][]" }
    ],
    name: "claim",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function"
  },
  {
    inputs: [
      { internalType: "address", name: "user", type: "address" },
      { internalType: "address", name: "token", type: "address" }
    ],
    name: "claimed",
    outputs: [{ internalType: "uint256", name: "", type: "uint256" }],
    stateMutability: "view",
    type: "function"
  }
];

// src/actions/merkl.ts
function prepareClaimMerklRewards(params) {
  const distributorAddress = MERKL_DISTRIBUTOR_ADDRESSES[import_chains2.base.id];
  if (!distributorAddress) {
    throw new Error("No Merkl distributor found for Base");
  }
  const { users, tokens, amounts, proofs } = params;
  if (users.length === 0) {
    throw new Error("No rewards to claim");
  }
  const data = (0, import_viem.encodeFunctionData)({
    abi: merklDistributorAbi,
    functionName: "claim",
    args: [users, tokens, amounts, proofs]
  });
  return { to: distributorAddress, data, value: 0n, chainId: import_chains2.base.id };
}

// src/actions/prepare.ts
var import_viem3 = require("viem");

// src/constants/abis.ts
var erc20Abi = [
  {
    type: "function",
    name: "name",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "string" }]
  },
  {
    type: "function",
    name: "symbol",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "string" }]
  },
  {
    type: "function",
    name: "decimals",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "uint8" }]
  },
  {
    type: "function",
    name: "totalSupply",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "balanceOf",
    stateMutability: "view",
    inputs: [{ name: "account", type: "address" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "allowance",
    stateMutability: "view",
    inputs: [
      { name: "owner", type: "address" },
      { name: "spender", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "approve",
    stateMutability: "nonpayable",
    inputs: [
      { name: "spender", type: "address" },
      { name: "amount", type: "uint256" }
    ],
    outputs: [{ type: "bool" }]
  },
  {
    type: "function",
    name: "transfer",
    stateMutability: "nonpayable",
    inputs: [
      { name: "to", type: "address" },
      { name: "amount", type: "uint256" }
    ],
    outputs: [{ type: "bool" }]
  },
  {
    type: "function",
    name: "transferFrom",
    stateMutability: "nonpayable",
    inputs: [
      { name: "from", type: "address" },
      { name: "to", type: "address" },
      { name: "amount", type: "uint256" }
    ],
    outputs: [{ type: "bool" }]
  },
  {
    type: "event",
    name: "Transfer",
    inputs: [
      { name: "from", type: "address", indexed: true },
      { name: "to", type: "address", indexed: true },
      { name: "value", type: "uint256", indexed: false }
    ]
  },
  {
    type: "event",
    name: "Approval",
    inputs: [
      { name: "owner", type: "address", indexed: true },
      { name: "spender", type: "address", indexed: true },
      { name: "value", type: "uint256", indexed: false }
    ]
  }
];
var erc4626Abi = [
  ...erc20Abi,
  {
    type: "function",
    name: "asset",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "address" }]
  },
  {
    type: "function",
    name: "totalAssets",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "convertToShares",
    stateMutability: "view",
    inputs: [{ name: "assets", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "convertToAssets",
    stateMutability: "view",
    inputs: [{ name: "shares", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "maxDeposit",
    stateMutability: "view",
    inputs: [{ name: "receiver", type: "address" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "previewDeposit",
    stateMutability: "view",
    inputs: [{ name: "assets", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "deposit",
    stateMutability: "nonpayable",
    inputs: [
      { name: "assets", type: "uint256" },
      { name: "receiver", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "maxMint",
    stateMutability: "view",
    inputs: [{ name: "receiver", type: "address" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "previewMint",
    stateMutability: "view",
    inputs: [{ name: "shares", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "mint",
    stateMutability: "nonpayable",
    inputs: [
      { name: "shares", type: "uint256" },
      { name: "receiver", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "maxWithdraw",
    stateMutability: "view",
    inputs: [{ name: "owner", type: "address" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "previewWithdraw",
    stateMutability: "view",
    inputs: [{ name: "assets", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "withdraw",
    stateMutability: "nonpayable",
    inputs: [
      { name: "assets", type: "uint256" },
      { name: "receiver", type: "address" },
      { name: "owner", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "maxRedeem",
    stateMutability: "view",
    inputs: [{ name: "owner", type: "address" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "previewRedeem",
    stateMutability: "view",
    inputs: [{ name: "shares", type: "uint256" }],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "redeem",
    stateMutability: "nonpayable",
    inputs: [
      { name: "shares", type: "uint256" },
      { name: "receiver", type: "address" },
      { name: "owner", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "event",
    name: "Deposit",
    inputs: [
      { name: "sender", type: "address", indexed: true },
      { name: "owner", type: "address", indexed: true },
      { name: "assets", type: "uint256", indexed: false },
      { name: "shares", type: "uint256", indexed: false }
    ]
  },
  {
    type: "event",
    name: "Withdraw",
    inputs: [
      { name: "sender", type: "address", indexed: true },
      { name: "receiver", type: "address", indexed: true },
      { name: "owner", type: "address", indexed: true },
      { name: "assets", type: "uint256", indexed: false },
      { name: "shares", type: "uint256", indexed: false }
    ]
  }
];
var pausableAbi = [
  {
    type: "function",
    name: "paused",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "bool" }]
  }
];
var yoGatewayAbi = [
  {
    type: "function",
    name: "deposit",
    stateMutability: "nonpayable",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "assets", type: "uint256" },
      { name: "minSharesOut", type: "uint256" },
      { name: "receiver", type: "address" },
      { name: "partnerId", type: "uint32" }
    ],
    outputs: [{ name: "sharesOut", type: "uint256" }]
  },
  {
    type: "function",
    name: "redeem",
    stateMutability: "nonpayable",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "shares", type: "uint256" },
      { name: "minAssetsOut", type: "uint256" },
      { name: "receiver", type: "address" },
      { name: "partnerId", type: "uint32" }
    ],
    outputs: [{ name: "assetsOrRequestId", type: "uint256" }]
  },
  {
    type: "function",
    name: "quotePreviewDeposit",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "assets", type: "uint256" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "quotePreviewRedeem",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "shares", type: "uint256" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "quotePreviewWithdraw",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "assets", type: "uint256" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "quoteConvertToAssets",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "shares", type: "uint256" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "quoteConvertToShares",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "assets", type: "uint256" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "getShareAllowance",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "owner", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "getAssetAllowance",
    stateMutability: "view",
    inputs: [
      { name: "yoVault", type: "address" },
      { name: "owner", type: "address" }
    ],
    outputs: [{ type: "uint256" }]
  },
  {
    type: "function",
    name: "registry",
    stateMutability: "view",
    inputs: [],
    outputs: [{ type: "address" }]
  },
  {
    type: "event",
    name: "YoGatewayDeposit",
    inputs: [
      { name: "partnerId", type: "uint32", indexed: true },
      { name: "yoVault", type: "address", indexed: true },
      { name: "sender", type: "address", indexed: true },
      { name: "receiver", type: "address", indexed: false },
      { name: "assets", type: "uint256", indexed: false },
      { name: "shares", type: "uint256", indexed: false }
    ]
  },
  {
    type: "event",
    name: "YoGatewayRedeem",
    inputs: [
      { name: "partnerId", type: "uint32", indexed: true },
      { name: "yoVault", type: "address", indexed: true },
      { name: "receiver", type: "address", indexed: true },
      { name: "shares", type: "uint256", indexed: false },
      { name: "assetsOrRequestId", type: "uint256", indexed: false },
      { name: "instant", type: "bool", indexed: false }
    ]
  }
];

// src/constants/addresses.ts
var YO_GATEWAY_ADDRESS = "0xF1EeE0957267b1A474323Ff9CfF7719E964969FA";
var VAULT_REGISTRY_ADDRESS = "0x56c3119DC3B1a75763C87D5B0A2C55E489502232";
var YO_ORACLE_ADDRESS = "0x6E879d0CcC85085A709eBf5539224f53d0D396B0";
var REDEEMER_ADDRESS = "0x0439e941841f97dc1334d1a433379c6fcdcc2162";
var VAULTS = {
  yoETH: {
    address: "0x3a43aec53490cb9fa922847385d82fe25d0e9de7",
    name: "YO ETH Vault",
    symbol: "yoETH",
    network: "base",
    underlying: {
      symbol: "WETH",
      decimals: 18,
      address: {
        1: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
        // Ethereum
        8453: "0x4200000000000000000000000000000000000006"
        // Base
      }
    },
    chains: [1, 8453]
  },
  yoBTC: {
    address: "0xbcbc8cb4d1e8ed048a6276a5e94a3e952660bcbc",
    name: "YO BTC Vault",
    symbol: "yoBTC",
    network: "base",
    underlying: {
      symbol: "cbBTC",
      decimals: 8,
      address: {
        1: "0xcbb7c0000ab88b473b1f5afd9ef808440eed33bf",
        8453: "0xcbb7c0000ab88b473b1f5afd9ef808440eed33bf"
      }
    },
    chains: [1, 8453]
  },
  yoUSD: {
    address: "0x0000000f2eb9f69274678c76222b35eec7588a65",
    name: "YO USD Vault",
    symbol: "yoUSD",
    network: "base",
    underlying: {
      symbol: "USDC",
      decimals: 6,
      address: {
        1: "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
        8453: "0x833589fcd6edb6e08f4c7c32d4f71b54bda02913",
        42161: "0xaf88d065e77c8cC2239327C5EDb3A432268e5831"
      }
    },
    chains: [1, 8453, 42161]
  },
  yoEUR: {
    address: "0x50c749ae210d3977adc824ae11f3c7fd10c871e9",
    name: "YO EUR Vault",
    symbol: "yoEUR",
    network: "base",
    underlying: {
      symbol: "EURC",
      decimals: 6,
      address: {
        1: "0x1aBaEA1f7C830bD89Acc67eC4af516284b1bC33c",
        8453: "0x60a3E35Cc302bFA44Cb288Bc5a4F316Fdb1adb42"
      }
    },
    chains: [1, 8453]
  },
  yoGOLD: {
    address: "0x586675A3a46B008d8408933cf42d8ff6c9CC61a1",
    name: "YO Gold Vault",
    symbol: "yoGOLD",
    network: "ethereum",
    underlying: {
      symbol: "XAUt",
      decimals: 6,
      address: {
        1: "0x68749665FF8D2d112Fa859AA293F07A622782F38"
      }
    },
    chains: [1]
  },
  yoUSDT: {
    address: "0xb9a7da9e90d3b428083bae04b860faa6325b721e",
    name: "YO USDT Vault",
    symbol: "yoUSDT",
    network: "ethereum",
    underlying: {
      symbol: "USDT",
      decimals: 6,
      address: {
        1: "0xdac17f958d2ee523a2206206994597c13d831ec7"
      }
    },
    chains: [1]
  }
};
function getVaultByAddress(address) {
  const normalizedAddress = address.toLowerCase();
  return Object.values(VAULTS).find((v) => v.address.toLowerCase() === normalizedAddress);
}
function getVaultsForChain(chainId) {
  return Object.values(VAULTS).filter((v) => v.chains.includes(chainId));
}
function getAllVaults() {
  return Object.values(VAULTS);
}

// src/utils/math.ts
var import_big = __toESM(require("big.js"), 1);
import_big.default.DP = 100;
import_big.default.NE = -100;
import_big.default.PE = 100;
function mul(a, b) {
  return toBig(a).times(toBig(b)).toString();
}
function mulDiv(a, b, c) {
  return toBig(a).times(toBig(b)).div(toBig(c)).toString();
}
function div(a, b) {
  return toBig(a).div(toBig(b)).toString();
}
function sub(a, b) {
  return toBig(a).sub(toBig(b)).toString();
}
function add(a, b) {
  return toBig(a).plus(toBig(b)).toString();
}
function gt(a, b) {
  return toBig(a).gt(toBig(b));
}
function gte(a, b) {
  return toBig(a).gte(toBig(b));
}
function lt(a, b) {
  return toBig(a).lt(toBig(b));
}
function lte(a, b) {
  return toBig(a).lte(toBig(b));
}
function eq(a, b) {
  return toBig(a).eq(toBig(b));
}
function toBig(value) {
  return new import_big.default(value.toString());
}
function toFixed(a, precision = 2, roundingMode = 3) {
  return new import_big.default(a.toString()).toFixed(precision, roundingMode);
}
function toNumber(a) {
  return new import_big.default(a.toString()).toNumber();
}
function toStr(a) {
  return new import_big.default(a.toString()).toString();
}
function abs(a) {
  return new import_big.default(a.toString()).abs().toString();
}
function pow(a, exp) {
  return new import_big.default(a.toString()).pow(exp).toString();
}
function sqrt(a) {
  return new import_big.default(a.toString()).sqrt().toString();
}
function toBigInt(a) {
  return BigInt(a.toString());
}
function applySlippage(value, slippageBps) {
  return value - value * BigInt(slippageBps) / 10000n;
}

// src/utils/native.ts
var NATIVE_TOKEN_ADDRESS = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE";
var ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
function isNativeToken(address) {
  const lower = address.toLowerCase();
  return lower === ZERO_ADDRESS || lower === NATIVE_TOKEN_ADDRESS.toLowerCase();
}

// src/utils/validation.ts
var import_viem2 = require("viem");
function validateAddress(address, label) {
  if (!(0, import_viem2.isAddress)(address)) {
    throw new Error(`[yo-kit] Invalid ${label} address: ${address}`);
  }
}
function validateAmount(amount, label) {
  if (amount <= 0n) {
    throw new Error(`[yo-kit] ${label} must be greater than 0`);
  }
}

// src/actions/prepare.ts
function prepareApprove(params) {
  const { token, spender = YO_GATEWAY_ADDRESS, amount } = params;
  const data = (0, import_viem3.encodeFunctionData)({
    abi: erc20Abi,
    functionName: "approve",
    args: [spender, amount]
  });
  return { to: token, data, value: 0n };
}
async function prepareDeposit(publicClient, params) {
  const { vault, amount, recipient, slippageBps = 50, partnerId = 9999 } = params;
  validateAddress(vault, "vault");
  validateAmount(amount, "amount");
  if (!recipient) {
    throw new Error("[yo-kit] recipient is required for prepareDeposit");
  }
  const expectedShares = await publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quotePreviewDeposit",
    args: [vault, amount]
  });
  const minShares = applySlippage(expectedShares, slippageBps);
  const data = (0, import_viem3.encodeFunctionData)({
    abi: yoGatewayAbi,
    functionName: "deposit",
    args: [vault, amount, minShares, recipient, partnerId]
  });
  return { to: YO_GATEWAY_ADDRESS, data, value: 0n };
}
async function prepareRedeem(publicClient, params) {
  const { vault, shares, recipient, slippageBps = 50, partnerId = 9999 } = params;
  validateAddress(vault, "vault");
  validateAmount(shares, "shares");
  if (!recipient) {
    throw new Error("[yo-kit] recipient is required for prepareRedeem");
  }
  const expectedAssets = await publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quotePreviewRedeem",
    args: [vault, shares]
  });
  const minAssetsOut = params.minAssetsOut ?? applySlippage(expectedAssets, slippageBps);
  const data = (0, import_viem3.encodeFunctionData)({
    abi: yoGatewayAbi,
    functionName: "redeem",
    args: [vault, shares, minAssetsOut, recipient, partnerId]
  });
  return { to: YO_GATEWAY_ADDRESS, data, value: 0n };
}
async function prepareDepositWithApproval(publicClient, params) {
  const { token, owner, amount } = params;
  const txs = [];
  if (!isNativeToken(token)) {
    const allowance = await publicClient.readContract({
      address: token,
      abi: erc20Abi,
      functionName: "allowance",
      args: [owner, YO_GATEWAY_ADDRESS]
    });
    if (allowance < amount) {
      txs.push(prepareApprove({ token, amount }));
    }
  }
  const depositTx = await prepareDeposit(publicClient, params);
  txs.push(depositTx);
  return txs;
}
async function prepareRedeemWithApproval(publicClient, params) {
  const { vault, owner, shares } = params;
  const txs = [];
  const allowance = await publicClient.readContract({
    address: vault,
    abi: erc20Abi,
    functionName: "allowance",
    args: [owner, YO_GATEWAY_ADDRESS]
  });
  if (allowance < shares) {
    txs.push(prepareApprove({ token: vault, amount: shares }));
  }
  const redeemTx = await prepareRedeem(publicClient, params);
  txs.push(redeemTx);
  return txs;
}

// src/actions/redeem.ts
var import_viem4 = require("viem");
async function waitForRedeemReceipt(publicClient, hash) {
  const receipt = await publicClient.waitForTransactionReceipt({ hash });
  for (const log of receipt.logs) {
    try {
      const event = (0, import_viem4.decodeEventLog)({
        abi: yoGatewayAbi,
        data: log.data,
        topics: log.topics
      });
      if (event.eventName === "YoGatewayRedeem") {
        const args = event.args;
        return {
          hash,
          status: receipt.status,
          instant: args.instant,
          assetsOrRequestId: args.assetsOrRequestId,
          shares: args.shares,
          blockNumber: receipt.blockNumber
        };
      }
    } catch {
    }
  }
  throw new Error(
    `YoGatewayRedeem event not found in transaction ${hash}. The transaction may have succeeded but the event could not be decoded.`
  );
}

// src/api/client.ts
var API_BASE_URL = "https://api.yo.xyz";
var ApiClient = class {
  constructor(config) {
    __publicField(this, "baseUrl");
    this.baseUrl = config?.baseUrl ?? API_BASE_URL;
  }
  async fetch(endpoint, options) {
    const url = `${this.baseUrl}${endpoint}`;
    const response = await fetch(url, {
      ...options,
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        ...options?.headers
      }
    });
    if (!response.ok) {
      throw new ApiError(response.status, `API request failed: ${response.statusText}`);
    }
    return response.json();
  }
  buildUrl(path, params) {
    const searchParams = new URLSearchParams();
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== void 0) {
          searchParams.append(key, String(value));
        }
      }
    }
    const queryString = searchParams.toString();
    return queryString ? `${path}?${queryString}` : path;
  }
};
var ApiError = class extends Error {
  constructor(statusCode, message) {
    super(message);
    this.statusCode = statusCode;
    this.name = "ApiError";
  }
};
function createApiClient(config) {
  return new ApiClient(config);
}

// src/api/merkl.ts
var import_viem5 = require("viem");
var import_chains3 = require("viem/chains");
var MerklApiClient = class {
  constructor(config) {
    __publicField(this, "baseUrl");
    this.baseUrl = config?.baseUrl ?? MERKL_API_BASE_URL;
  }
  async fetch(endpoint, params) {
    const url = new URL(`${this.baseUrl}${endpoint}`);
    if (params) {
      for (const [key, value] of Object.entries(params)) {
        if (value !== void 0 && value !== null) {
          url.searchParams.append(key, String(value));
        }
      }
    }
    const response = await fetch(url.toString(), {
      headers: {
        Accept: "application/json"
      }
    });
    if (!response.ok) {
      throw new Error(`Merkl API request failed: ${response.statusText}`);
    }
    return response.json();
  }
  async getCampaigns(options) {
    const PAGE_SIZE = 50;
    let page = 0;
    const allCampaigns = [];
    while (true) {
      const params = {
        creatorAddress: options?.creatorAddress ?? MERKL_CREATOR_ADDRESS,
        withOpportunity: true,
        items: PAGE_SIZE,
        page
      };
      if (options?.status) {
        params.status = options.status;
      }
      const campaigns = await this.fetch("/campaigns", params);
      allCampaigns.push(...campaigns);
      if (campaigns.length < PAGE_SIZE) {
        break;
      }
      page++;
    }
    return allCampaigns;
  }
  async getUserRewards(userAddress, chainId = import_chains3.base.id) {
    const params = {
      chainIds: chainId,
      reloadChainId: chainId
    };
    const rewards = await this.fetch(
      `/users/${userAddress}/rewards/breakdowns`,
      params
    );
    const chainRewards = rewards[0];
    if (!chainRewards) return null;
    const yoRewards = chainRewards.rewards.filter(
      (reward) => (0, import_viem5.getAddress)(reward.token.address) === (0, import_viem5.getAddress)(YO_TOKEN.address)
    );
    if (yoRewards.length === 0) return null;
    return { ...chainRewards, rewards: yoRewards };
  }
};
function createMerklApiClient(config) {
  return new MerklApiClient(config);
}
function getClaimableAmount(reward) {
  const amount = BigInt(reward.amount);
  const claimed = BigInt(reward.claimed);
  return amount > claimed ? amount - claimed : 0n;
}
function hasClaimableRewards(chainRewards) {
  return chainRewards.rewards.some((reward) => getClaimableAmount(reward) > 0n);
}
function prepareClaimParams(userAddress, chainRewards) {
  const users = [];
  const tokens = [];
  const amounts = [];
  const proofs = [];
  for (const reward of chainRewards.rewards) {
    const claimable = getClaimableAmount(reward);
    if (claimable > 0n && reward.proofs && reward.proofs.length > 0) {
      users.push(userAddress);
      tokens.push((0, import_viem5.getAddress)(reward.token.address));
      amounts.push(BigInt(reward.amount));
      proofs.push(reward.proofs);
    }
  }
  return { users, tokens, amounts, proofs };
}
function getTotalClaimable(chainRewards) {
  return chainRewards.rewards.reduce((total, reward) => total + getClaimableAmount(reward), 0n);
}
var VAULT_ADDRESSES = new Set(Object.values(VAULTS).map((v) => v.address.toLowerCase()));
var VAULT_ADDRESS_TO_SYMBOL = new Map(
  Object.values(VAULTS).map((v) => [v.address.toLowerCase(), v.symbol])
);
var VAULT_ASSET_PATTERNS = {
  yoETH: /yoETH|^ETH$|WETH|stETH|wstETH|cbETH|rETH|weETH|frxETH|oETH/i,
  yoUSD: /yoUSD|^USD$|USDC|USDT|DAI|USDS|gtUSDC/i,
  yoEUR: /yoEUR|^EUR$|EURC/i,
  yoBTC: /yoBTC|^BTC$|WBTC|cbBTC|LBTC/i,
  yoGOLD: /yoGOLD|^GOLD$|XAUT/i
};
function resolveVaultType(opportunity) {
  if (opportunity.identifier) {
    const match = VAULT_ADDRESS_TO_SYMBOL.get(opportunity.identifier.toLowerCase());
    if (match) return match;
  }
  const tokens = opportunity.tokens ?? [];
  for (const token of tokens) {
    const match = VAULT_ADDRESS_TO_SYMBOL.get(token.address.toLowerCase());
    if (match) return match;
  }
  for (const token of tokens) {
    for (const [vaultName, pattern] of Object.entries(VAULT_ASSET_PATTERNS)) {
      if (pattern.test(token.symbol)) return vaultName;
    }
  }
  return null;
}
function getPendingRewardSets(chainRewards) {
  const opportunityIds = /* @__PURE__ */ new Set();
  const vaultTypes = /* @__PURE__ */ new Set();
  if (!chainRewards?.rewards) return { opportunityIds, vaultTypes };
  for (const reward of chainRewards.rewards) {
    if (!reward.breakdowns) continue;
    for (const breakdown of reward.breakdowns) {
      const claimable = BigInt(breakdown.amount) - BigInt(breakdown.claimed);
      if (claimable <= 0n) continue;
      if (breakdown.opportunity?.id) {
        opportunityIds.add(breakdown.opportunity.id);
      }
      const vt = resolveVaultType(breakdown.opportunity);
      if (vt) {
        vaultTypes.add(vt);
      }
    }
  }
  return { opportunityIds, vaultTypes };
}
function campaignHasPendingRewards(campaign, pendingRewards) {
  const opp = campaign.Opportunity;
  if (!opp) return false;
  const isPrimaryYoVault = opp.identifier && VAULT_ADDRESSES.has(opp.identifier.toLowerCase());
  if (isPrimaryYoVault) {
    return pendingRewards.vaultTypes.has(resolveVaultType(opp) ?? "");
  }
  return pendingRewards.opportunityIds.has(opp.id ?? "");
}

// src/api/leaderboard.ts
async function getWeeklyRewards(client, tokenAddress) {
  const response = await client.fetch(
    `/api/v1/rewards/weekly/${tokenAddress}`
  );
  return response.data;
}
async function getAllTimeRewards(client, tokenAddress) {
  const response = await client.fetch(
    `/api/v1/rewards/all-time/${tokenAddress}`
  );
  return response.data;
}
async function getUserRewardsByAsset(client, userAddress, tokenAddress = YO_TOKEN.address) {
  const response = await client.fetch(
    `/api/v1/rewards/user/${userAddress}/${tokenAddress}`
  );
  return response.data;
}

// src/constants/pricing.ts
var DEFAULT_ASSET_IDS = [
  "ethereum",
  "bitcoin",
  "usd-coin",
  "tether",
  "euro-coin",
  "tether-gold",
  "coinbase-wrapped-btc",
  "wrapped-bitcoin",
  "wrapped-steth",
  "rocket-pool-eth",
  "coinbase-wrapped-staked-eth"
];

// src/api/pricing.ts
async function getPrices(client) {
  const response = await client.fetch("/api/v1/pricing/usd", {
    method: "POST",
    body: JSON.stringify({ ids: DEFAULT_ASSET_IDS })
  });
  return response.data;
}

// src/api/user.ts
async function getUserTransactionHistory(client, network, vaultAddress, userAddress, limit) {
  const endpoint = client.buildUrl(
    `/api/v1/history/user/${network}/${vaultAddress}/${userAddress}`,
    { limit }
  );
  const response = await client.fetch(endpoint);
  return response.data;
}
async function getUserPendingRedemptions(client, network, vaultAddress, userAddress) {
  const response = await client.fetch(
    `/api/v1/vault/pending-redeems/${network}/${vaultAddress}/${userAddress}`
  );
  return response.data;
}
async function getUserPerformance(client, network, vaultAddress, userAddress) {
  const response = await client.fetch(
    `/api/v1/performance/user/${network}/${vaultAddress}/${userAddress}`
  );
  return response.data;
}
async function getUserSnapshots(client, network, vaultAddress, userAddress) {
  const response = await client.fetch(
    `/api/v1/performance/timeseries/${network}/${vaultAddress}/${userAddress}`
  );
  return response.data;
}
async function getUserBalances(client, userAddress) {
  const response = await client.fetch(
    `/api/v1/user/balance/${userAddress}`
  );
  return response.data;
}
async function getUserPoints(client, userAddress) {
  try {
    const response = await client.fetch(
      `/api/v1/users/wallet/${userAddress}/points`
    );
    return response.data;
  } catch (error) {
    if (error instanceof ApiError && error.statusCode === 404) {
      return null;
    }
    throw error;
  }
}

// src/api/vault.ts
function normalizeTimestamp(ts) {
  return ts > 4102444800 ? Math.floor(ts / 1e3) : ts;
}
async function getVaultSnapshot(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getVaultPendingRedeems(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/pending-redeems/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getVaultYieldTimeseries(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/yield/timeseries/${network}/${vaultAddress}`
  );
  return response.data.map((point) => ({
    timestamp: normalizeTimestamp(point.timestamp),
    value: parseFloat(point.yield) / 100
    // Convert percentage to decimal
  }));
}
async function getVaultTvlTimeseries(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/tvl/timeseries/${network}/${vaultAddress}`
  );
  return response.data.map((point) => ({
    timestamp: normalizeTimestamp(point.timestamp),
    value: parseFloat(point.tvl)
  }));
}
async function getVaults(client, options) {
  const endpoint = client.buildUrl("/api/v1/vault/stats", {
    secondary: options?.secondary
  });
  const response = await client.fetch(endpoint);
  return response.data;
}
async function getVaultPercentile(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/yield/percentile/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getVaultHistory(client, network, vaultAddress, limit) {
  const endpoint = client.buildUrl(`/api/v1/history/vault/${network}/${vaultAddress}`, { limit });
  const response = await client.fetch(endpoint);
  return response.data;
}
async function getVaultAllocationsTimeSeries(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/allocations/timeseries/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getSharePriceHistory(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/price/timeseries/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getVaultHistoryAllNetworks(client, vaultAddress, options) {
  const endpoint = client.buildUrl("/api/v1/history/vaults", {
    vaultAddress,
    cursor: options?.cursor,
    limit: options?.limit,
    whale: options?.filters?.whale,
    type: options?.filters?.type,
    period: options?.filters?.period,
    network: options?.filters?.network
  });
  const response = await client.fetch(endpoint);
  return response.data;
}
async function getGlobalVaultHistory(client, options) {
  const endpoint = client.buildUrl("/api/v1/history/vaults", {
    cursor: options?.cursor,
    limit: options?.limit,
    whale: options?.filters?.whale,
    type: options?.filters?.type,
    period: options?.filters?.period,
    network: options?.filters?.network
  });
  const response = await client.fetch(endpoint);
  return response.data;
}
async function getTotalTvlTimeseries(client) {
  const response = await client.fetch(
    "/api/v1/vault/tvl/timeseries/total"
  );
  return response.data;
}
async function getVaultPerformance(client, network, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/performance/vault/${network}/${vaultAddress}`
  );
  return response.data;
}
async function getPerformanceBenchmark(client, vaultAddress) {
  const response = await client.fetch(
    `/api/v1/vault/performance-benchmark/${vaultAddress}`
  );
  return response.data;
}

// src/client.ts
var import_viem6 = require("viem");

// src/constants/chains.ts
var import_chains4 = require("viem/chains");
var SUPPORTED_CHAINS = {
  [import_chains4.mainnet.id]: import_chains4.mainnet,
  [import_chains4.base.id]: import_chains4.base,
  [import_chains4.arbitrum.id]: import_chains4.arbitrum
};
var SUPPORTED_CHAIN_IDS = Object.keys(SUPPORTED_CHAINS).map(Number);
function isSupportedChain(chainId) {
  return chainId in SUPPORTED_CHAINS;
}
function getChain(chainId) {
  return SUPPORTED_CHAINS[chainId];
}
var CHAIN_ID_TO_NETWORK = {
  1: "ethereum",
  8453: "base",
  42161: "arbitrum"
};
var NETWORK_TO_CHAIN_ID = {
  ethereum: 1,
  base: 8453,
  arbitrum: 42161
};

// src/reads/balance.ts
async function getTokenBalance(publicClient, token, account) {
  const [balance, decimals] = await Promise.all([
    publicClient.readContract({
      address: token,
      abi: erc20Abi,
      functionName: "balanceOf",
      args: [account]
    }),
    publicClient.readContract({
      address: token,
      abi: erc20Abi,
      functionName: "decimals"
    })
  ]);
  return {
    token,
    balance,
    decimals
  };
}
async function getShareBalance(publicClient, vaultAddress, account) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "balanceOf",
    args: [account]
  });
}
async function getUserVaultPosition(publicClient, vaultAddress, account) {
  const shares = await publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "balanceOf",
    args: [account]
  });
  let assets = 0n;
  if (shares > 0n) {
    assets = await publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "previewRedeem",
      args: [shares]
    });
  }
  return {
    shares,
    assets
  };
}
async function getUserPositionsAllChains(publicClients, account, vaults) {
  if (!vaults.length) return [];
  const primaryCallsByChain = /* @__PURE__ */ new Map();
  const secondaryCallsByChain = /* @__PURE__ */ new Map();
  for (let i = 0; i < vaults.length; i++) {
    const v = vaults[i];
    const chainId = v.chain.id;
    if (!primaryCallsByChain.has(chainId)) primaryCallsByChain.set(chainId, []);
    primaryCallsByChain.get(chainId).push({
      vaultIndex: i,
      contract: {
        address: v.contracts.vaultAddress,
        abi: erc4626Abi,
        functionName: "balanceOf",
        args: [account]
      }
    });
    const secondaries = v.secondaryVaults;
    if (secondaries?.length) {
      for (const sv of secondaries) {
        if (!publicClients[sv.chain.id]) continue;
        if (!secondaryCallsByChain.has(sv.chain.id)) secondaryCallsByChain.set(sv.chain.id, []);
        secondaryCallsByChain.get(sv.chain.id).push({
          vaultIndex: i,
          contract: {
            address: sv.shareAsset.address,
            abi: erc20Abi,
            functionName: "balanceOf",
            args: [account]
          }
        });
      }
    }
  }
  const primaryShares = new Array(vaults.length).fill(0n);
  const secondaryShares = new Array(vaults.length).fill(0n);
  const multicallPromises = [];
  for (const [chainId, calls] of primaryCallsByChain) {
    const client = publicClients[chainId];
    if (!client) continue;
    multicallPromises.push(
      client.multicall({ contracts: calls.map((c) => c.contract) }).then((results) => {
        for (let j = 0; j < results.length; j++) {
          const r = results[j];
          if (r.status === "success") {
            primaryShares[calls[j].vaultIndex] = r.result;
          }
        }
      }).catch(() => {
      })
    );
  }
  for (const [chainId, calls] of secondaryCallsByChain) {
    const client = publicClients[chainId];
    if (!client) continue;
    multicallPromises.push(
      client.multicall({ contracts: calls.map((c) => c.contract) }).then((results) => {
        for (let j = 0; j < results.length; j++) {
          const r = results[j];
          const idx = calls[j].vaultIndex;
          if (r.status === "success") {
            secondaryShares[idx] = (secondaryShares[idx] ?? 0n) + r.result;
          }
        }
      }).catch(() => {
      })
    );
  }
  await Promise.all(multicallPromises);
  const convertCallsByChain = /* @__PURE__ */ new Map();
  for (let i = 0; i < vaults.length; i++) {
    const totalShares = primaryShares[i] + secondaryShares[i];
    if (totalShares === 0n) continue;
    const v = vaults[i];
    const chainId = v.chain.id;
    if (!convertCallsByChain.has(chainId)) convertCallsByChain.set(chainId, []);
    if (primaryShares[i] > 0n) {
      convertCallsByChain.get(chainId).push({
        vaultIndex: i,
        contract: {
          address: v.contracts.vaultAddress,
          abi: erc4626Abi,
          functionName: "previewRedeem",
          args: [primaryShares[i]]
        }
      });
    }
    if (secondaryShares[i] > 0n) {
      convertCallsByChain.get(chainId).push({
        vaultIndex: i,
        contract: {
          address: v.contracts.vaultAddress,
          abi: erc4626Abi,
          functionName: "previewRedeem",
          args: [secondaryShares[i]]
        }
      });
    }
  }
  const assets = new Array(vaults.length).fill(0n);
  const convertPromises = [];
  for (const [chainId, calls] of convertCallsByChain) {
    const client = publicClients[chainId];
    if (!client) continue;
    convertPromises.push(
      client.multicall({ contracts: calls.map((c) => c.contract) }).then((results) => {
        for (let j = 0; j < results.length; j++) {
          const r = results[j];
          const idx = calls[j].vaultIndex;
          if (r.status === "success") {
            assets[idx] = (assets[idx] ?? 0n) + r.result;
          }
        }
      }).catch(() => {
      })
    );
  }
  await Promise.all(convertPromises);
  const positions = [];
  for (let i = 0; i < vaults.length; i++) {
    const totalShares = primaryShares[i] + secondaryShares[i];
    if (totalShares > 0n) {
      positions.push({
        vault: vaults[i],
        position: { shares: totalShares, assets: assets[i] }
      });
    }
  }
  return positions;
}

// src/reads/gateway.ts
async function quotePreviewDeposit(publicClient, vaultAddress, assets) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quotePreviewDeposit",
    args: [vaultAddress, assets]
  });
}
async function quotePreviewRedeem(publicClient, vaultAddress, shares) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quotePreviewRedeem",
    args: [vaultAddress, shares]
  });
}
async function quotePreviewWithdraw(publicClient, vaultAddress, assets) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quotePreviewWithdraw",
    args: [vaultAddress, assets]
  });
}
async function quoteConvertToAssets(publicClient, vaultAddress, shares) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quoteConvertToAssets",
    args: [vaultAddress, shares]
  });
}
async function quoteConvertToShares(publicClient, vaultAddress, assets) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "quoteConvertToShares",
    args: [vaultAddress, assets]
  });
}
async function getShareAllowance(publicClient, vaultAddress, owner) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "getShareAllowance",
    args: [vaultAddress, owner]
  });
}
async function getAssetAllowance(publicClient, vaultAddress, owner) {
  return publicClient.readContract({
    address: YO_GATEWAY_ADDRESS,
    abi: yoGatewayAbi,
    functionName: "getAssetAllowance",
    args: [vaultAddress, owner]
  });
}

// src/reads/vault.ts
async function getVaultState(publicClient, vaultAddress) {
  const [name, symbol, decimals, totalSupply, totalAssets, asset] = await Promise.all([
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "name"
    }),
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "symbol"
    }),
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "decimals"
    }),
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "totalSupply"
    }),
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "totalAssets"
    }),
    publicClient.readContract({
      address: vaultAddress,
      abi: erc4626Abi,
      functionName: "asset"
    })
  ]);
  const assetDecimals = await publicClient.readContract({
    address: asset,
    abi: erc20Abi,
    functionName: "decimals"
  });
  const exchangeRate = totalSupply > 0n ? await publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "convertToAssets",
    args: [10n ** BigInt(decimals)]
  }) : 10n ** BigInt(assetDecimals);
  return {
    address: vaultAddress,
    name,
    symbol,
    decimals,
    totalAssets,
    totalSupply,
    asset,
    assetDecimals,
    exchangeRate
  };
}
async function previewDeposit(publicClient, vaultAddress, assets) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "previewDeposit",
    args: [assets]
  });
}
async function previewRedeem(publicClient, vaultAddress, shares) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "previewRedeem",
    args: [shares]
  });
}
async function convertToShares(publicClient, vaultAddress, assets) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "convertToShares",
    args: [assets]
  });
}
async function convertToAssets(publicClient, vaultAddress, shares) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "convertToAssets",
    args: [shares]
  });
}
async function maxDeposit(publicClient, vaultAddress, receiver) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "maxDeposit",
    args: [receiver]
  });
}
async function maxRedeem(publicClient, vaultAddress, owner) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "maxRedeem",
    args: [owner]
  });
}
async function isPaused(publicClient, vaultAddress) {
  return publicClient.readContract({
    address: vaultAddress,
    abi: pausableAbi,
    functionName: "paused"
  });
}
async function getIdleBalance(publicClient, vaultAddress) {
  const asset = await publicClient.readContract({
    address: vaultAddress,
    abi: erc4626Abi,
    functionName: "asset"
  });
  return publicClient.readContract({
    address: asset,
    abi: erc20Abi,
    functionName: "balanceOf",
    args: [vaultAddress]
  });
}

// src/services/vault-read.service.ts
var VaultReadService = class {
  constructor(publicClient) {
    this.publicClient = publicClient;
  }
  getVaultState(vaultAddress) {
    return getVaultState(this.publicClient, vaultAddress);
  }
  previewDeposit(vaultAddress, assets) {
    return previewDeposit(this.publicClient, vaultAddress, assets);
  }
  previewRedeem(vaultAddress, shares) {
    return previewRedeem(this.publicClient, vaultAddress, shares);
  }
  convertToAssets(vaultAddress, shares) {
    return convertToAssets(this.publicClient, vaultAddress, shares);
  }
  convertToShares(vaultAddress, assets) {
    return convertToShares(this.publicClient, vaultAddress, assets);
  }
  isPaused(vaultAddress) {
    return isPaused(this.publicClient, vaultAddress);
  }
  getIdleBalance(vaultAddress) {
    return getIdleBalance(this.publicClient, vaultAddress);
  }
  quotePreviewDeposit(vaultAddress, assets) {
    return quotePreviewDeposit(this.publicClient, vaultAddress, assets);
  }
  quotePreviewRedeem(vaultAddress, shares) {
    return quotePreviewRedeem(this.publicClient, vaultAddress, shares);
  }
  quotePreviewWithdraw(vaultAddress, assets) {
    return quotePreviewWithdraw(this.publicClient, vaultAddress, assets);
  }
  quoteConvertToAssets(vaultAddress, shares) {
    return quoteConvertToAssets(this.publicClient, vaultAddress, shares);
  }
  quoteConvertToShares(vaultAddress, assets) {
    return quoteConvertToShares(this.publicClient, vaultAddress, assets);
  }
  getShareAllowance(vaultAddress, owner) {
    return getShareAllowance(this.publicClient, vaultAddress, owner);
  }
  getAssetAllowance(vaultAddress, owner) {
    return getAssetAllowance(this.publicClient, vaultAddress, owner);
  }
};

// src/reads/allowance.ts
var MAX_UINT256 = 2n ** 256n - 1n;
async function getTokenAllowance(publicClient, token, owner, spender) {
  if (isNativeToken(token)) {
    console.log("[getTokenAllowance] native token detected \u2014 returning MAX_UINT256");
    return { token, owner, spender, allowance: MAX_UINT256 };
  }
  const allowance = await publicClient.readContract({
    address: token,
    abi: erc20Abi,
    functionName: "allowance",
    args: [owner, spender]
  });
  return {
    token,
    owner,
    spender,
    allowance
  };
}
async function hasEnoughAllowance(publicClient, token, owner, spender, amount) {
  const { allowance } = await getTokenAllowance(publicClient, token, owner, spender);
  return allowance >= amount;
}

// src/services/user-read.service.ts
var UserReadService = class {
  constructor(publicClient) {
    this.publicClient = publicClient;
  }
  getTokenBalance(token, account) {
    return getTokenBalance(this.publicClient, token, account);
  }
  getShareBalance(vaultAddress, account) {
    return getShareBalance(this.publicClient, vaultAddress, account);
  }
  getUserPosition(vaultAddress, account) {
    return getUserVaultPosition(this.publicClient, vaultAddress, account);
  }
  getAllowance(token, owner, spender) {
    return getTokenAllowance(this.publicClient, token, owner, spender);
  }
  hasEnoughAllowance(token, owner, spender, amount) {
    return hasEnoughAllowance(this.publicClient, token, owner, spender, amount);
  }
};

// src/services/transaction.service.ts
var TransactionService = class {
  constructor(publicClients, defaultChainId, partnerId) {
    this.publicClients = publicClients;
    this.defaultChainId = defaultChainId;
    this.partnerId = partnerId;
  }
  getClient(chainId) {
    return this.publicClients[chainId] ?? this.publicClients[this.defaultChainId];
  }
  // Prepared transactions
  prepareApprove(params) {
    return prepareApprove(params);
  }
  async prepareDeposit(params) {
    return prepareDeposit(this.getClient(params.chainId), {
      ...params,
      partnerId: params.partnerId ?? this.partnerId
    });
  }
  async prepareRedeem(params) {
    return prepareRedeem(this.getClient(params.chainId), {
      ...params,
      partnerId: params.partnerId ?? this.partnerId
    });
  }
  async prepareDepositWithApproval(params) {
    return prepareDepositWithApproval(this.getClient(params.chainId), {
      ...params,
      partnerId: params.partnerId ?? this.partnerId
    });
  }
  async prepareRedeemWithApproval(params) {
    return prepareRedeemWithApproval(this.getClient(params.chainId), {
      ...params,
      partnerId: params.partnerId ?? this.partnerId
    });
  }
  // Confirmation
  async waitForTransaction(hash, chainId) {
    const receipt = await this.getClient(chainId).waitForTransactionReceipt({ hash });
    return {
      hash,
      status: receipt.status,
      blockNumber: receipt.blockNumber,
      gasUsed: receipt.gasUsed
    };
  }
  async waitForRedeemReceipt(hash, chainId) {
    return waitForRedeemReceipt(this.getClient(chainId), hash);
  }
};

// src/services/vault-api.service.ts
var VaultApiService = class {
  constructor(apiClient, chainId) {
    this.apiClient = apiClient;
    __publicField(this, "network");
    this.network = CHAIN_ID_TO_NETWORK[chainId];
  }
  // Vault
  getVaults(options) {
    return getVaults(this.apiClient, options);
  }
  getVaultSnapshot(vaultAddress) {
    return getVaultSnapshot(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getVaultHistory(vaultAddress, limit) {
    return getVaultHistory(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress, limit);
  }
  getVaultHistoryAllNetworks(vaultAddress, options) {
    return getVaultHistoryAllNetworks(this.apiClient, vaultAddress, options);
  }
  getGlobalVaultHistory(options) {
    return getGlobalVaultHistory(this.apiClient, options);
  }
  getVaultYieldHistory(vaultAddress) {
    return getVaultYieldTimeseries(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getVaultTvlHistory(vaultAddress) {
    return getVaultTvlTimeseries(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getTotalTvlTimeseries() {
    return getTotalTvlTimeseries(this.apiClient);
  }
  getVaultPercentile(vaultAddress) {
    return getVaultPercentile(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getSharePriceHistory(vaultAddress) {
    return getSharePriceHistory(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getVaultAllocationsTimeSeries(vaultAddress) {
    return getVaultAllocationsTimeSeries(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getVaultPendingRedeems(vaultAddress) {
    return getVaultPendingRedeems(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getVaultPerformance(vaultAddress) {
    return getVaultPerformance(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress);
  }
  getPerformanceBenchmark(vaultAddress) {
    return getPerformanceBenchmark(this.apiClient, vaultAddress);
  }
  // User
  getUserHistory(vaultAddress, user, limit) {
    return getUserTransactionHistory(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress, user, limit);
  }
  getUserPerformance(vaultAddress, user) {
    return getUserPerformance(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress, user);
  }
  getPendingRedemptions(vaultAddress, user) {
    return getUserPendingRedemptions(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress, user);
  }
  getUserSnapshots(vaultAddress, user) {
    return getUserSnapshots(this.apiClient, this.resolveVaultNetwork(vaultAddress), vaultAddress, user);
  }
  getUserBalances(user) {
    return getUserBalances(this.apiClient, user);
  }
  // Rewards leaderboard
  getWeeklyRewards(tokenAddress) {
    return getWeeklyRewards(this.apiClient, tokenAddress);
  }
  getAllTimeRewards(tokenAddress) {
    return getAllTimeRewards(this.apiClient, tokenAddress);
  }
  getUserRewardsByAsset(userAddress, tokenAddress) {
    return getUserRewardsByAsset(this.apiClient, userAddress, tokenAddress);
  }
  // Pricing
  getPrices() {
    return getPrices(this.apiClient);
  }
  resolveVaultNetwork(vaultAddress) {
    const config = getVaultByAddress(vaultAddress);
    if (!config) return this.network;
    return config.network;
  }
};

// src/services/merkl.service.ts
var import_chains6 = require("viem/chains");

// src/reads/merkl.ts
async function getClaimedAmount(publicClient, userAddress, tokenAddress, chainId) {
  const distributorAddress = MERKL_DISTRIBUTOR_ADDRESSES[chainId];
  if (!distributorAddress) {
    throw new Error(`No Merkl distributor found for chain ${chainId}`);
  }
  return publicClient.readContract({
    address: distributorAddress,
    abi: merklDistributorAbi,
    functionName: "claimed",
    args: [userAddress, tokenAddress]
  });
}
async function getClaimedAmounts(publicClient, userAddress, tokenAddresses, chainId) {
  const distributorAddress = MERKL_DISTRIBUTOR_ADDRESSES[chainId];
  if (!distributorAddress) {
    throw new Error(`No Merkl distributor found for chain ${chainId}`);
  }
  const results = await Promise.all(
    tokenAddresses.map(
      (token) => publicClient.readContract({
        address: distributorAddress,
        abi: merklDistributorAbi,
        functionName: "claimed",
        args: [userAddress, token]
      })
    )
  );
  const claimedMap = /* @__PURE__ */ new Map();
  tokenAddresses.forEach((token, index) => {
    claimedMap.set(token, results[index] ?? 0n);
  });
  return claimedMap;
}

// src/services/merkl.service.ts
var MerklService = class {
  constructor(publicClients, merklApiClient) {
    this.merklApiClient = merklApiClient;
    /** Rewards are always distributed on Base — use the Base public client for on-chain reads */
    __publicField(this, "basePublicClient");
    this.basePublicClient = publicClients[import_chains6.base.id];
  }
  // API
  getCampaigns(options) {
    return this.merklApiClient.getCampaigns(options);
  }
  getRewards(user) {
    return this.merklApiClient.getUserRewards(user, import_chains6.base.id);
  }
  // On-chain
  getClaimedAmount(user, token) {
    return getClaimedAmount(this.basePublicClient, user, token, import_chains6.base.id);
  }
  // Hybrid
  async getClaimableRewards(user) {
    const apiRewards = await this.getRewards(user);
    if (!apiRewards || apiRewards.rewards.length === 0) return null;
    const tokenAddresses = apiRewards.rewards.map((r) => r.token.address);
    const claimedMap = await getClaimedAmounts(
      this.basePublicClient,
      user,
      tokenAddresses,
      import_chains6.base.id
    );
    return {
      ...apiRewards,
      rewards: apiRewards.rewards.map((reward) => ({
        ...reward,
        claimed: (claimedMap.get(reward.token.address) ?? 0n).toString()
      }))
    };
  }
  // Prepare
  prepareClaimRewards(userAddress, chainRewards) {
    const claimParams = prepareClaimParams(userAddress, chainRewards);
    if (claimParams.users.length === 0) {
      throw new Error("No claimable rewards");
    }
    return prepareClaimMerklRewards(claimParams);
  }
  // Utilities
  getClaimableAmount(reward) {
    return getClaimableAmount(reward);
  }
  hasClaimableRewards(chainRewards) {
    return hasClaimableRewards(chainRewards);
  }
  getTotalClaimable(chainRewards) {
    return getTotalClaimable(chainRewards);
  }
  /**
   * Fetch claimable rewards and return which opportunities/vault types have pending rewards.
   * Use the result with `campaignHasPendingRewards()` to check individual campaigns.
   */
  async getPendingRewards(user) {
    const rewards = await this.getClaimableRewards(user);
    return getPendingRewardSets(rewards);
  }
  getPendingRewardSets(chainRewards) {
    return getPendingRewardSets(chainRewards);
  }
};

// src/client.ts
var YoClient = class {
  constructor(config) {
    __publicField(this, "chainId");
    __publicField(this, "publicClient");
    __publicField(this, "publicClients");
    __publicField(this, "apiClient");
    __publicField(this, "merklApiClient");
    __publicField(this, "partnerId");
    __publicField(this, "vaultReads");
    __publicField(this, "userReads");
    __publicField(this, "transactions");
    __publicField(this, "vaultApi");
    __publicField(this, "merkl");
    this.chainId = config.chainId;
    this.partnerId = config.partnerId ?? 9999;
    const clients = {};
    for (const cid of SUPPORTED_CHAIN_IDS) {
      if (config.publicClients?.[cid]) {
        clients[cid] = config.publicClients[cid];
      } else {
        clients[cid] = (0, import_viem6.createPublicClient)({
          chain: SUPPORTED_CHAINS[cid],
          transport: (0, import_viem6.http)()
        });
      }
    }
    this.publicClients = clients;
    this.publicClient = clients[config.chainId];
    this.apiClient = createApiClient();
    this.merklApiClient = createMerklApiClient();
    this.vaultReads = new VaultReadService(this.publicClient);
    this.userReads = new UserReadService(this.publicClient);
    this.transactions = new TransactionService(this.publicClients, this.chainId, this.partnerId);
    this.vaultApi = new VaultApiService(this.apiClient, this.chainId);
    this.merkl = new MerklService(this.publicClients, this.merklApiClient);
  }
  get network() {
    return CHAIN_ID_TO_NETWORK[this.chainId];
  }
  getVaults(options) {
    return this.vaultApi.getVaults(options);
  }
  // Vault reads
  getVaultState(vaultAddress) {
    return this.vaultReads.getVaultState(vaultAddress);
  }
  previewDeposit(vaultAddress, assets) {
    return this.vaultReads.previewDeposit(vaultAddress, assets);
  }
  previewRedeem(vaultAddress, shares) {
    return this.vaultReads.previewRedeem(vaultAddress, shares);
  }
  convertToAssets(vaultAddress, shares) {
    return this.vaultReads.convertToAssets(vaultAddress, shares);
  }
  convertToShares(vaultAddress, assets) {
    return this.vaultReads.convertToShares(vaultAddress, assets);
  }
  isPaused(vaultAddress) {
    return this.vaultReads.isPaused(vaultAddress);
  }
  getIdleBalance(vaultAddress) {
    return this.vaultReads.getIdleBalance(vaultAddress);
  }
  // User reads
  getTokenBalance(token, account) {
    return this.userReads.getTokenBalance(token, account);
  }
  getShareBalance(vaultAddress, account) {
    return this.userReads.getShareBalance(vaultAddress, account);
  }
  getUserPosition(vaultAddress, account) {
    return this.userReads.getUserPosition(vaultAddress, account);
  }
  getAllowance(token, owner, spender) {
    return this.userReads.getAllowance(token, owner, spender);
  }
  hasEnoughAllowance(token, owner, spender, amount) {
    return this.userReads.hasEnoughAllowance(token, owner, spender, amount);
  }
  /**
   * Fetch user positions across ALL chains using multicall.
   * Aggregates primary + secondary vault share balances.
   *
   * If `vaults` is provided, uses that data directly.
   * Otherwise fetches vault stats from the API first.
   */
  async getUserPositionsAllChains(account, vaults) {
    const vaultData = vaults ?? await this.getVaults();
    return getUserPositionsAllChains(this.publicClients, account, vaultData);
  }
  // Transaction confirmation
  waitForTransaction(hash, chainId) {
    return this.transactions.waitForTransaction(hash, chainId);
  }
  waitForRedeemReceipt(hash, chainId) {
    return this.transactions.waitForRedeemReceipt(hash, chainId);
  }
  // Prepared transactions
  prepareApprove(params) {
    return this.transactions.prepareApprove(params);
  }
  prepareDeposit(params) {
    return this.transactions.prepareDeposit(params);
  }
  prepareRedeem(params) {
    return this.transactions.prepareRedeem(params);
  }
  prepareDepositWithApproval(params) {
    return this.transactions.prepareDepositWithApproval(params);
  }
  prepareRedeemWithApproval(params) {
    return this.transactions.prepareRedeemWithApproval(params);
  }
  // Gateway quote helpers
  quotePreviewDeposit(vaultAddress, assets) {
    return this.vaultReads.quotePreviewDeposit(vaultAddress, assets);
  }
  quotePreviewRedeem(vaultAddress, shares) {
    return this.vaultReads.quotePreviewRedeem(vaultAddress, shares);
  }
  quotePreviewWithdraw(vaultAddress, assets) {
    return this.vaultReads.quotePreviewWithdraw(vaultAddress, assets);
  }
  quoteConvertToAssets(vaultAddress, shares) {
    return this.vaultReads.quoteConvertToAssets(vaultAddress, shares);
  }
  quoteConvertToShares(vaultAddress, assets) {
    return this.vaultReads.quoteConvertToShares(vaultAddress, assets);
  }
  // Gateway allowance helpers
  getShareAllowance(vaultAddress, owner) {
    return this.vaultReads.getShareAllowance(vaultAddress, owner);
  }
  getAssetAllowance(vaultAddress, owner) {
    return this.vaultReads.getAssetAllowance(vaultAddress, owner);
  }
  // Rewards leaderboard
  getWeeklyRewards(tokenAddress) {
    return this.vaultApi.getWeeklyRewards(tokenAddress);
  }
  getAllTimeRewards(tokenAddress) {
    return this.vaultApi.getAllTimeRewards(tokenAddress);
  }
  // API methods
  getPrices() {
    return this.vaultApi.getPrices();
  }
  getVaultHistory(vaultAddress, limit) {
    return this.vaultApi.getVaultHistory(vaultAddress, limit);
  }
  getVaultHistoryAllNetworks(vaultAddress, options) {
    return this.vaultApi.getVaultHistoryAllNetworks(vaultAddress, options);
  }
  /**
   * Fetch deposit/withdrawal history across ALL vaults and networks in a single API call.
   */
  getGlobalVaultHistory(options) {
    return this.vaultApi.getGlobalVaultHistory(options);
  }
  getVaultSnapshot(vaultAddress) {
    return this.vaultApi.getVaultSnapshot(vaultAddress);
  }
  getVaultYieldHistory(vaultAddress) {
    return this.vaultApi.getVaultYieldHistory(vaultAddress);
  }
  getVaultTvlHistory(vaultAddress) {
    return this.vaultApi.getVaultTvlHistory(vaultAddress);
  }
  getTotalTvlTimeseries() {
    return this.vaultApi.getTotalTvlTimeseries();
  }
  getVaultPercentile(vaultAddress) {
    return this.vaultApi.getVaultPercentile(vaultAddress);
  }
  getSharePriceHistory(vaultAddress) {
    return this.vaultApi.getSharePriceHistory(vaultAddress);
  }
  getVaultAllocationsTimeSeries(vaultAddress) {
    return this.vaultApi.getVaultAllocationsTimeSeries(vaultAddress);
  }
  getUserHistory(vaultAddress, user, limit) {
    return this.vaultApi.getUserHistory(vaultAddress, user, limit);
  }
  getUserPerformance(vaultAddress, user) {
    return this.vaultApi.getUserPerformance(vaultAddress, user);
  }
  getPendingRedemptions(vaultAddress, user) {
    return this.vaultApi.getPendingRedemptions(vaultAddress, user);
  }
  getVaultPendingRedeems(vaultAddress) {
    return this.vaultApi.getVaultPendingRedeems(vaultAddress);
  }
  getVaultPerformance(vaultAddress) {
    return this.vaultApi.getVaultPerformance(vaultAddress);
  }
  getPerformanceBenchmark(vaultAddress) {
    return this.vaultApi.getPerformanceBenchmark(vaultAddress);
  }
  getUserSnapshots(vaultAddress, user) {
    return this.vaultApi.getUserSnapshots(vaultAddress, user);
  }
  getUserBalances(user) {
    return this.vaultApi.getUserBalances(user);
  }
  getUserRewardsByAsset(user, tokenAddress) {
    return this.vaultApi.getUserRewardsByAsset(user, tokenAddress);
  }
  // Merkl reward methods
  getMerklCampaigns(options) {
    return this.merkl.getCampaigns(options);
  }
  getMerklRewards(user) {
    return this.merkl.getRewards(user);
  }
  getMerklClaimedAmount(user, token) {
    return this.merkl.getClaimedAmount(user, token);
  }
  getClaimableRewards(user) {
    return this.merkl.getClaimableRewards(user);
  }
  prepareClaimMerklRewards(userAddress, chainRewards) {
    return this.merkl.prepareClaimRewards(userAddress, chainRewards);
  }
  getMerklClaimableAmount(reward) {
    return this.merkl.getClaimableAmount(reward);
  }
  hasMerklClaimableRewards(chainRewards) {
    return this.merkl.hasClaimableRewards(chainRewards);
  }
  getMerklTotalClaimable(chainRewards) {
    return this.merkl.getTotalClaimable(chainRewards);
  }
  /**
   * Fetch which campaigns/opportunities have pending (claimable) rewards for a user.
   * Returns sets of opportunity IDs and vault types for easy matching against campaigns.
   *
   * Usage:
   * ```ts
   * const pending = await client.getMerklPendingRewards(userAddress)
   * const hasPending = campaignHasPendingRewards(campaign, pending)
   * ```
   */
  getMerklPendingRewards(user) {
    return this.merkl.getPendingRewards(user);
  }
  /**
   * Extract pending reward sets from already-fetched chain rewards.
   * Use this when you've already called `getClaimableRewards()`.
   */
  getMerklPendingRewardSets(chainRewards) {
    return this.merkl.getPendingRewardSets(chainRewards);
  }
};
function createYoClient(config) {
  return new YoClient(config);
}

// src/utils/format.ts
var import_viem7 = require("viem");
function formatTokenAmount(amount, decimals, options) {
  const { maxDecimals = 6, trimTrailingZeros = true } = options ?? {};
  const formatted = (0, import_viem7.formatUnits)(amount, decimals);
  const parts = formatted.split(".");
  const whole = parts[0] ?? "0";
  const fraction = parts[1] ?? "";
  let trimmedFraction = fraction.slice(0, maxDecimals);
  if (trimTrailingZeros) {
    trimmedFraction = trimmedFraction.replace(/0+$/, "");
  }
  if (!trimmedFraction) {
    return whole;
  }
  return `${whole}.${trimmedFraction}`;
}
function parseTokenAmount(amount, decimals) {
  return (0, import_viem7.parseUnits)(amount, decimals);
}
function formatUsd(amount) {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
    maximumFractionDigits: 2
  }).format(amount);
}
function formatPercent(value, decimals = 2) {
  return `${(value * 100).toFixed(decimals)}%`;
}
function formatCompactNumber(value) {
  return new Intl.NumberFormat("en-US", {
    notation: "compact",
    compactDisplay: "short",
    maximumFractionDigits: 2
  }).format(value);
}
